<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Ghader Mirzaghaderi" />


<title>qPCR data analysis</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>







<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">qPCR data analysis</h1>
<h4 class="author">Ghader Mirzaghaderi</h4>



<div id="qpcr-data-analysis" class="section level1">
<h1>qPCR data analysis</h1>
<p>Real-time polymerase chain reaction (real-time PCR) is widely used in
biological studies. Various analysis methods are employed on the
real-time PCR data to measure the mRNA levels under different
experimental conditions.</p>
<p>‘rtpcr’ package was developed for amplification efficiency
calculation, statistical analysis and bar plot representation of
real-time PCR data in R. By accounting for up to two reference genes and
amplification efficiency values, a general calculation methodology
described by <a href="https://doi.org/10.1186/s12859-017-1949-5">Ganger
et al. (2017)</a> and
<a href="https://doi.org/10.1016/j.tibtech.2018.12.002">Taylor et
al. (2019)</a>, matching both
<a href="https://doi.org/10.1006/meth.2001.1262">Livak and Schmittgen
(2001)</a> and <a href="https://doi.org/10.1093/nar/30.9.e36">Pfaffl et
al. (2002) </a> methods was used. Based on the experimental conditions,
the functions of the ‘rtpcr’ package use t-test (for experiments with a
two-level factor), analysis of variance, analysis of covariance (ANCOVA)
or analysis of repeated measure data to calculate the fold change (FC,
<span class="math inline">\({\Delta\Delta C_t}\)</span> method) or
relative expression (RE, <span class="math inline">\({\Delta
C_t}\)</span> method). The functions further provide standard errors and
confidence interval for means, apply statistical mean comparisons and
present significance. To facilitate function application, different data
sets were used as examples and the outputs were explained. An
outstanding feature of ‘rtpcr’ package is providing publication-ready
bar plots with various controlling arguments for experiments with up to
three different factors which are further editable by ggplot2
functions.</p>
<p># Calculations The basic method for estimating gene expression
between conditions relies on calculating fold differences using the PCR
amplification efficiency (<span class="math inline">\(E\)</span>) and
the threshold cycle (also called the crossing point, <span class="math inline">\(C_t\)</span>). Among the various approaches for
analyzing real-time PCR data, the Livak method—also known as the <span class="math inline">\(2^{-\Delta\Delta C_t}\)</span> method—is widely
used for its simplicity and reliability. In this method, the fold change
(FC) expression, <span class="math inline">\(2^{-\Delta\Delta
C_t}\)</span>, in a Treatment (Tr) group relative to a Control (Co)
group is calculated as follows:</p>
<p><span class="math display">\[
\begin{align*}
\text{Fold change}
&amp;= 2^{-\Delta\Delta C_t} \\
&amp;= \frac{2^{-(C_{t,\text{target}} - C_{t,\text{ref}})_{\text{Tr}}}}
        {2^{-(C_{t,\text{target}} - C_{t,\text{ref}})_{\text{Co}}}} \\
&amp;= 2^{-[(C_{t,\text{target}} - C_{t,\text{ref}})_{\text{Tr}}
       - (C_{t,\text{target}} - C_{t,\text{ref}})_{\text{Co}}]} \\
&amp;= 2^{-[(\Delta C_t)_{\text{Tr}} - (\Delta C_t)_{\text{Co}}]}
\end{align*}
\]</span></p>
<pre><code>Here, $\Delta C_t$ is the difference between target Ct and reference Ct values for a given sample. Livak method assumes that both the target and reference genes are amplified with efficiencies close to 100%, allowing for the relative quantification of gene expression levels.</code></pre>
<p>On the other hand, the Pfaffl method offers a more flexible approach
by accounting for differences in amplification efficiencies between the
target and reference genes. This method adjusts the calculated
expression ratio by incorporating the specific amplification
efficiencies, thus providing a more accurate representation of the
relative gene expression levels.</p>
<p><span class="math display">\[
\text{Fold change} =
\frac{E_{\text{target}}^{-(C_{t,\text{Tr}} - C_{t,\text{Co}})}}
     {E_{\text{ref}}^{-(C_{t,\text{Tr}} - C_{t,\text{Co}})}}
\]</span></p>
<pre><code># A generalized calculation method
The `rtpcr` package was developed for the R environment in the major operating systems. The package functions are mainly based on the calculation of efficiency-weighted $\Delta C_t$ $(w\Delta C_t)$ values from target and reference gene Ct (equation 3). $w\Delta C_t$  values are weighted for the amplification efficiencies as described by Ganger et al. (2017) except that log2 is used instead of log10:</code></pre>
<p><span class="math display">\[
w\Delta C_t =
\log_2(E_{\text{target}}) \cdot C_{t,\text{target}}
-
\log_2(E_{\text{ref}}) \cdot C_{t,\text{ref}}
\]</span></p>
<pre><code>The relative expression of the target gene normalized to that of reference gene(s) within the same sample or condition is called relative expression (RE). From the mean $w\Delta C_t$ values over biological replicates, RE of a target gene can be calculated for each condition according to the equation </code></pre>
<p><span class="math display">\[
\text{Relative Expression} = 2^{-\overline{w\Delta C_t}}
\]</span></p>
<pre><code>Relative expression is only calibrated for the reference gene(s) and not for a control condition. However, often one condition is considered as calibrator and the fold change (FC) expression in other conditions is calculated relative to the calibrator. Examples are Treatment versus Control where Control is served as the calibrator, or time 0 versus time 1 (e.g. after 1 hour) and time 2 (e.g. after 2 hours) where time 0 is served as the reference or calibrator level. So, calibrator is the reference level or sample that all others are compared to. The fold change (FC) expression of a target gene for the reference or calibrator level is 1 because it is not changed compared to itself. The fold change expression of a target gene due to the treatment can be calculated as follows: </code></pre>
<p><span class="math display">\[
\text{Fold Change due to Treatment} =
2^{-(\overline{w\Delta C_t}_{\text{Tr}} - \overline{w\Delta
C_t}_{\text{Co}})}
\]</span></p>
<pre><code>Standard error of the FC and RE means is calculated according to &lt;a href=&quot;https://doi.org/10.1016/j.tibtech.2018.12.002&quot;&gt;Taylor et al. (2019)&lt;/a&gt; in `rtpcr` package.</code></pre>
<p>Here, a brief methodology is presented but details about the <span class="math inline">\(w\Delta C_t\)</span> calculations and statistical
analysis are available in
<a href="https://doi.org/10.1186/s12859-017-1949-5">Ganger et
al. (2017)</a>. Importantly, because relative expression values follow a
lognormal distribution, a normal distribution is expected for the <span class="math inline">\(w \Delta C_t\)</span> values making it possible to
apply t-test or analysis of variance. Following analysis, <span class="math inline">\(w\Delta C_t\)</span> values are statistically
compared and standard deviations and confidence interval are calculated,
but the transformation <span class="math inline">\(y = 2^{-x}\)</span>
is applied in the final step in order to report the results.</p>
<p># References Livak, Kenneth J, and Thomas D Schmittgen. 2001.
Analysis of Relative Gene Expression Data Using Real-Time Quantitative
PCR and the Double Delta CT Method. Methods 25 (4).
<a href="https://doi.org/10.1006/meth.2001.1262">doi.org/10.1006/meth.2001.1262</a>.</p>
<p>Ganger, MT, Dietz GD, Ewing SJ. 2017. A common base method for
analysis of qPCR data and the application of simple blocking in qPCR
experiments. BMC bioinformatics 18, 1-11.
<a href="https://doi.org/10.1186/s12859-017-1949-5">doi.org/10.1186/s12859-017-1949-5</a>.</p>
<p>Pfaffl MW, Horgan GW, Dempfle L. 2002. Relative expression software
tool (REST©) for group-wise comparison and statistical analysis of
relative expression results in real-time PCR. Nucleic acids research 30,
e36-e36.
<a href="https://doi.org/10.1093/nar/30.9.e36">doi.org/10.1093/nar/30.9.e36</a>.</p>
<p>Taylor SC, Nadeau K, Abbasi M, Lachance C, Nguyen M, Fenrich, J.
2019. The ultimate qPCR experiment: producing publication quality,
reproducible data the first time. Trends in Biotechnology, 37(7),
761-774.
<a href="https://doi.org/10.1016/j.tibtech.2018.12.002">doi.org/10.1016/j.tibtech.2018.12.002</a>.</p>
<p>Yuan, JS, Ann Reed, Feng Chen, and Neal Stewart. 2006. Statistical
Analysis of Real-Time PCR Data. BMC Bioinformatics 7 (85).
<a href="https://doi.org/10.1186/1471-2105-7-85">doi.org/10.1186/1471-2105-7-85</a>.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
