<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>qPCR Data Analysis • rtpcr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" sizes="any" href="favicon.ico">
<link rel="manifest" href="site.webmanifest">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="qPCR Data Analysis">
<meta name="description" content="Tools for qPCR data analysis using Delta Ct and Delta Delta Ct methods, including t-test, Wilcoxon-test, ANOVA models, and publication-ready visualizations. The package supports multiple target, and multiple reference genes, and uses a calculation framework adopted from Ganger et al. (2017) &lt;doi:10.1186/s12859-017-1949-5&gt; and Taylor et al. (2019) &lt;doi:10.1016/j.tibtech.2018.12.002&gt;, covering both the Livak and Pfaffl methods.">
<meta property="og:description" content="Tools for qPCR data analysis using Delta Ct and Delta Delta Ct methods, including t-test, Wilcoxon-test, ANOVA models, and publication-ready visualizations. The package supports multiple target, and multiple reference genes, and uses a calculation framework adopted from Ganger et al. (2017) &lt;doi:10.1186/s12859-017-1949-5&gt; and Taylor et al. (2019) &lt;doi:10.1016/j.tibtech.2018.12.002&gt;, covering both the Livak and Pfaffl methods.">
<meta property="og:image" content="https://mirzaghaderi.github.io/rtpcr/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">rtpcr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/Manual.html">Manual</a></li>
    <li><a class="dropdown-item" href="articles/Method.html">Calculation method</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mirzaghaderi/rtpcr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="rtpcr-">rtpcr <a class="anchor" aria-label="anchor" href="#rtpcr-"></a>
</h1>
</div>
<!-- badges: start -->

<p>⁠<!-- badges: end --></p>
<p>rtpcr is a tool for analysis of RT-qPCR gene expression data using <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Δ</mi><mi>C</mi><mi>t</mi></mrow><annotation encoding="application/x-tex">\Delta Ct</annotation></semantics></math> and <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Δ</mi><mi>Δ</mi><mi>C</mi><mi>t</mi></mrow><annotation encoding="application/x-tex">\Delta\Delta Ct</annotation></semantics></math> methods, including t-tests and ANOVA, repeated-measures models, and publication-ready visualizations. The package implements a general calculation method adopted from Ganger et al. (2017) and Taylor et al. (2019), covering both the Livak and Pfaffl methods.</p>
<ul>
<li><a href="#functions">Functions</a></li>
<li>
<a href="#quick-start">Quick start</a>
<ul>
<li><a href="#installing-and-loading">Installing and loading</a></li>
</ul>
</li>
<li><a href="#input-data-structure">Input data structure</a></li>
<li><a href="#handling-missing-data">Handling missing data</a></li>
<li>
<a href="#data-analysis">Data analysis</a>
<ul>
<li><a href="#amplification-efficiency">Amplification efficiency</a></li>
<li><a href="#relative-expression">Relative expression</a></li>
</ul>
</li>
<li>
<a href="#output">Output</a>
<ul>
<li><a href="#data-output">Data output</a></li>
<li><a href="#plot-output">Plot output</a></li>
</ul>
</li>
<li><a href="#post-hoc-analysis">Post-hoc analysis</a></li>
<li><a href="#checking-normality-of-residuals">Checking normality of residuals</a></li>
<li><a href="#mean-of-technical-replicates">Mean of technical replicates</a></li>
<li><a href="#contact">Contact</a></li>
<li><a href="#citation">Citation</a></li>
<li><a href="#references">References</a></li>
</ul>
</div>
<div class="section level1">
<h1 id="functions">Functions<a class="anchor" aria-label="anchor" href="#functions"></a>
</h1>
<p>The rtpcr package gets efficiency (E) the Ct values of genes and performs different analyses using the following functions.</p>
<table class="table">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>ANOVA_DCt</code></td>
<td>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Δ</mi><mi>C</mi><mi>t</mi></mrow><annotation encoding="application/x-tex">\Delta Ct</annotation></semantics></math> ANOVA analysis</td>
</tr>
<tr class="even">
<td><code>ANOVA_DDCt</code></td>
<td>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Δ</mi><mi>Δ</mi><mi>C</mi><mi>t</mi></mrow><annotation encoding="application/x-tex">\Delta\Delta Ct</annotation></semantics></math> ANOVA analysis</td>
</tr>
<tr class="odd">
<td><code>TTEST_DDCt</code></td>
<td>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Δ</mi><mi>Δ</mi><mi>C</mi><mi>t</mi></mrow><annotation encoding="application/x-tex">\Delta\Delta Ct</annotation></semantics></math> method <em>t</em>-test analysis</td>
</tr>
<tr class="even">
<td><code>WILCOX_DDCt</code></td>
<td>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Δ</mi><mi>Δ</mi><mi>C</mi><mi>t</mi></mrow><annotation encoding="application/x-tex">\Delta\Delta Ct</annotation></semantics></math> method wilcox.test analysis</td>
</tr>
<tr class="odd">
<td><code>plotFactor</code></td>
<td>Bar plot of gene expression for one-, two- or three-factor experiments</td>
</tr>
<tr class="even">
<td><code>plotSingleGene</code></td>
<td>Creates a bar plot of relative gene expression (fold change) values from single gene analysis showing all pairwise significances.</td>
</tr>
<tr class="odd">
<td><code>Means_DDCt</code></td>
<td>Pairwise comparison of RE values for any user-specified effect</td>
</tr>
<tr class="even">
<td><code>efficiency</code></td>
<td>Amplification efficiency statistics and standard curves</td>
</tr>
<tr class="odd">
<td><code>meanTech</code></td>
<td>Calculate mean of technical replicates</td>
</tr>
<tr class="even">
<td><code>multiplot</code></td>
<td>Combine multiple ggplot objects into a single layout</td>
</tr>
<tr class="odd">
<td><code>compute_wDCt</code></td>
<td>Cleaning data and weighted delta Ct (wDCt) calculation</td>
</tr>
<tr class="even">
<td><code>long_to_wide</code></td>
<td>Converts a 4-column qPCR long data format to wide format</td>
</tr>
</tbody>
</table>
</div>
<div class="section level1">
<h1 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h1>
<div class="section level3">
<h3 id="installing-and-loading">Installing and loading<a class="anchor" aria-label="anchor" href="#installing-and-loading"></a>
</h3>
<p>The <code>rtpcr</code> package can be installed by running the following code in R:</p>
<p>from CRAN:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Installing from CRAN</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"rtpcr"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Loading the package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mirzaghaderi.github.io/rtpcr/">rtpcr</a></span><span class="op">)</span></span></code></pre></div>
<p>Or from from GitHub (developing version):</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/remotes/man/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"mirzaghaderi/rtpcr"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level1">
<h1 id="input-data-structure">Input data structure<a class="anchor" aria-label="anchor" href="#input-data-structure"></a>
</h1>
<p>For relative expression analysis (using <code>TTEST_DDCt</code>, <code>WILCOX_DDCt</code>, <code>ANOVA_DCt</code>, and <code>ANOVA_DDCt</code> functions), the amplification efficiency (<code>E</code>) and <code>Ct</code> or <code>Cq</code> values (the mean of technical replicates) is used for the input table. If the <code>E</code> values are not available you should use ‘2’ instead representing the complete primer amplification efficiency. The input data table should include the following columns from left to wright:</p>
<ol style="list-style-type: decimal">
<li>Experimental condition columns (and one block if available <a href="#note-1">NOTE 1</a>)</li>
<li>Replicates information (biological replicates or subjects; see <a href="#note-2">NOTE 2</a>, and <a href="#note-3">NOTE 3</a>)<br>
</li>
<li>Target genes efficiency and Ct values (a pair column for each gene).</li>
<li>Reference genes efficiency and Ct values (a pair column for each gene) <a href="#note-4">NOTE 4</a>.</li>
</ol>
<p>The package supports <strong>one or more target or reference gene(s)</strong>, supplied as efficiency–Ct column pairs. Reference gene columns must always appear last. Two sample input data sets are presented below.</p>
<figure><img src="reference/figures/sampleData1.png" class="center" style="width:100.0%" alt="Figure 1: A sample input data with one experimetal factor, replicate column and E/Ct information of target and reference genes"><figcaption aria-hidden="true">
Figure 1: A sample input data with one experimetal factor, replicate column and E/Ct information of target and reference genes
</figcaption></figure><p>If there is no blocking factor, the corresponding block columns should be omitted. However, a column for biological replicates (which may be named “Rep”, “id” or similar) is always required.</p>
<figure><img src="reference/figures/dataStructure1.png" class="center" style="width:100.0%" alt="Figure 2: A sample input data with two experimetal factors, blocking factor, replicate column and E/Ct information of target and reference genes"><figcaption aria-hidden="true">
Figure 2: A sample input data with two experimetal factors, blocking factor, replicate column and E/Ct information of target and reference genes
</figcaption></figure><div class="section level4">
<h4 id="note-1">NOTE 1<a class="anchor" aria-label="anchor" href="#note-1"></a>
</h4>
<p>When a qPCR experiment is done in multiple qPCR plates, variation resulting from the plates may interfere with the actual amount of gene expression. One solution is to conduct each plate as a randomized block so that at least one replicate of each treatment and control is present on a plate. Block effect is usually considered as random and its interaction with any main effect is not considered.</p>
</div>
<div class="section level4">
<h4 id="note-2">NOTE 2<a class="anchor" aria-label="anchor" href="#note-2"></a>
</h4>
<p>For <code>TTEST_DDCt</code> and <code>WILCOX_DDCt</code> (independent samples), <code>ANOVA_DCt</code>, and <code>ANOVA_DDCt</code> each row is from a separate and unique biological replicate. For example, a data frame with 12 rows has come from an experiment with 12 individuals. The repeated measure models are intended for experiments with repeated observations (e.g. time-course data). In repeated measure experiments the Replicate column contains identifiers for each individual (id or subject). For example, all rows with a <code>1</code> at Rep column correspond to a single individual, all rows with a <code>2</code> correspond to another individual, and so on, which have been sampled at specific time points.</p>
</div>
<div class="section level4">
<h4 id="note-3">NOTE 3<a class="anchor" aria-label="anchor" href="#note-3"></a>
</h4>
<p>Your data table may also include a column of technical replicates (For example, using one target and one reference genes, if you want to have 4 biological and 3 technical replicates under Control and Treatment conditions, there would be a table of 24 rows containing both biological replicates and technical replicate columns in the data). In this case, the <code>meanTech</code> function should be applied first to calculate the mean of the technical replicates. The resulting collapsed table is then used as the input for expression analysis. To use the <code>meanTech</code> function correctly, the technical replicate column must appear immediately after the biological replicate column (see <a href="#mean-of-technical-replicates">Mean of technical replicates</a> for an example).</p>
<p><img src="reference/figures/base.png" class="center" style="width:100.0%"></p>
</div>
<div class="section level4">
<h4 id="note-4">NOTE 4<a class="anchor" aria-label="anchor" href="#note-4"></a>
</h4>
<p>Complete amplification efficiency (E) in the rtpcr package is denoted by 2. This means that 2 indicates 100%, and 1.85 and 1.70 indicate 0.85% and 0.70% amplification efficiencies.</p>
</div>
</div>
<div class="section level1">
<h1 id="handling-missing-data">Handling missing data<a class="anchor" aria-label="anchor" href="#handling-missing-data"></a>
</h1>
<p>Missing Ct values for target genes is Handled using the <code>set_missing_target_Ct_to_40</code> function. If <code>TRUE</code>, missing target gene Ct values become 40; if <code>FALSE</code> (default), they become NA. missing Ct values of reference genes are always converted to NA. If there are more than one reference gene, NA in the place of the E or the Ct value of cause skipping that gene and remaining references are geometrically averaged in that replicate.</p>
</div>
<div class="section level1">
<h1 id="data-analysis">Data Analysis<a class="anchor" aria-label="anchor" href="#data-analysis"></a>
</h1>
<div class="section level3">
<h3 id="amplification-efficiency">Amplification Efficiency<a class="anchor" aria-label="anchor" href="#amplification-efficiency"></a>
</h3>
<p>The <code>efficiency</code> function calculates the amplification efficiency (E), slope, and R² statistics for genes, and performs pairwise comparisons of slopes. It takes a data frame in which the first column contains the dilution ratios, followed by the Ct value columns for each gene.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Applying the efficiency function</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"data_efficiency.csv"</span>, package <span class="op">=</span> <span class="st">"rtpcr"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">data</span></span>
<span><span class="co"># dilutions Gene1   Gene2   Gene3</span></span>
<span><span class="co"># 1.00  25.58   24.25   22.61</span></span>
<span><span class="co"># 1.00  25.54   24.13   22.68</span></span>
<span><span class="co"># 1.00  25.50   24.04   22.63</span></span>
<span><span class="co"># 0.50  26.71   25.56   23.67</span></span>
<span><span class="co"># 0.50  26.73   25.43   23.65</span></span>
<span><span class="co"># 0.50  26.87   26.01   23.70</span></span>
<span><span class="co"># 0.20  28.17   27.37   25.11</span></span>
<span><span class="co"># 0.20  28.07   26.94   25.12</span></span>
<span><span class="co"># 0.20  28.11   27.14   25.11</span></span>
<span><span class="co"># 0.10  29.20   28.05   26.17</span></span>
<span><span class="co"># 0.10  29.49   28.89   26.15</span></span>
<span><span class="co"># 0.10  29.07   28.32   26.15</span></span>
<span><span class="co"># 0.05  30.17   29.50   27.12</span></span>
<span><span class="co"># 0.05  30.14   29.93   27.14</span></span>
<span><span class="co"># 0.05  30.12   29.71   27.16</span></span>
<span><span class="co"># 0.02  31.35   30.69   28.52</span></span>
<span><span class="co"># 0.02  31.35   30.54   28.57</span></span>
<span><span class="co"># 0.02  31.35   30.04   28.53</span></span>
<span><span class="co"># 0.01  32.55   31.12   29.49</span></span>
<span><span class="co"># 0.01  32.45   31.29   29.48</span></span>
<span><span class="co"># 0.01  32.28   31.15   29.26</span></span>
<span></span>
<span><span class="co"># Analysis</span></span>
<span><span class="fu"><a href="reference/efficiency.html">efficiency</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># $Efficiency</span></span>
<span><span class="co">#    Gene     Slope        R2        E</span></span>
<span><span class="co"># 1 Gene1 -3.388094 0.9965504 1.973110</span></span>
<span><span class="co"># 2 Gene2 -3.528125 0.9713914 1.920599</span></span>
<span><span class="co"># 3 Gene3 -3.414551 0.9990278 1.962747</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># $Slope_compare</span></span>
<span><span class="co"># $contrasts</span></span>
<span><span class="co">#  contrast          estimate    SE df t.ratio p.value</span></span>
<span><span class="co">#  C2H2.26 - C2H2.01   0.1400 0.121 57   1.157  0.4837</span></span>
<span><span class="co">#  C2H2.26 - GAPDH     0.0265 0.121 57   0.219  0.9740</span></span>
<span><span class="co">#  C2H2.01 - GAPDH    -0.1136 0.121 57  -0.938  0.6186</span></span></code></pre></div>
<p><img src="reference/figures/standCur.png" class="center" style="width:67.0%"></p>
</div>
<div class="section level3">
<h3 id="relative-expression">Relative expression<a class="anchor" aria-label="anchor" href="#relative-expression"></a>
</h3>
<p><strong>Single factor experiment with two levels (e.g. Control and Treatment):</strong> <code><a href="reference/TTEST_DDCt.html">TTEST_DDCt()</a></code> function is used for relative expression analysis in treatment condition compared to control condition. Both paired and unpaired experimental designs are supported. if the data doesn’t follow t.test assumptions, the <code><a href="reference/WILCOX_DDCt.html">WILCOX_DDCt()</a></code> function can be used instead.</p>
<p><strong>Single factor experiment with more than two levels, or multi-factor experiments:</strong> In these cases, <code><a href="reference/ANOVA_DDCt.html">ANOVA_DDCt()</a></code> and <code><a href="reference/ANOVA_DCt.html">ANOVA_DCt()</a></code> functions are used for the analysis of qPCR data. By default, statistical analysis is performed based on uni- or multi-factorial Completely Randomized Design (CRD) or Randomized Complete Block Design (RCBD) design based on <code>numOfFactors</code> and the availability of <code>block</code>. However, optional custom model formula as a character string can be supplied to these functions. If provided, this overrides the default formula (uni- or multi-factorial CRD or RCBD design). The formula uses <code>wDCt</code> as the response variable (wDCt is automatically created by the function). For mixed models, include random effects using <code>lmer</code> syntax (e.g., <code>wDCt ~ Treatment + (1 | id)</code>). Below are a sample of most common models that can be used.</p>
<table class="table">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr class="header">
<th>Samples models may be used in <code><a href="reference/ANOVA_DCt.html">ANOVA_DCt()</a></code> or <code><a href="reference/ANOVA_DDCt.html">ANOVA_DDCt()</a></code> functions</th>
<th>Experimental design</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>wDCt ~ Condition</td>
<td>Completely Randomized Design (CRD). Can also be used for t.test with two independent samples. (<strong>default</strong>)</td>
</tr>
<tr class="even">
<td>wDCt ~ Factor1 * Factor2 * Factor3</td>
<td>Factorial under Completely Randomized Design (RCBD) (<strong>default</strong>)</td>
</tr>
<tr class="odd">
<td>wDCt ~ block + Factor1 * Factor2</td>
<td>Factorial under Randomized Complete Block Design (<strong>default</strong>)</td>
</tr>
<tr class="even">
<td>wDCt ~ time + (1 | id)</td>
<td>Repeated measure analysis: different time points. Also can be used for t.test with two paired samples.</td>
</tr>
<tr class="odd">
<td>wDCt ~ Condition * time + (1 | id)</td>
<td>Repeated measure analysis: split-plot in time</td>
</tr>
<tr class="even">
<td>wDCt ~ wDCt ~ Condition * time + (1 | block) + (1 | id)</td>
<td>Repeated measure analysis: split-plot in time</td>
</tr>
<tr class="odd">
<td>wDCt ~ Type + Concentration</td>
<td>Analysis of Covariance: Type is covariate</td>
</tr>
<tr class="even">
<td>wDCt ~ block + Type + Concentration</td>
<td>Analysis of Covariance with blocking factor: block and Type are covariates</td>
</tr>
</tbody>
</table>
<div class="section level4">
<h4 id="note">NOTE<a class="anchor" aria-label="anchor" href="#note"></a>
</h4>
<p>For CRD, RCBD, and factorial experiments arranged in either CRD or RCBD designs, you do not need to explicitly define a model. The package automatically selects an appropriate model based on the provided arguments. If no model is specified, the default model used is printed along with the output expression table.</p>
</div>
<div class="section level4">
<h4 id="note-5">NOTE<a class="anchor" aria-label="anchor" href="#note-5"></a>
</h4>
<p>Sometime samples are independent or paired (Repeated measure experiments). Examples: 1) Analyzing gene expression in different time points, or before and after treatment in each biological replicate; 2) Analyzing gene expression between two tissue types within the same organism. For such analysis types, if there are only two time points, we can use the <code>TTEST_DDCt</code> with the argument <code>paired = TRUE</code>; or <code>ANOVA_DDCt</code> (if there are two or more time points) with a repeated measure model such as <code>wDCt ~ Treatment + ( 1 | id)</code> or <code>wDCt ~ Treatment + ( 1 | Rep)</code>.</p>
<p><img src="reference/figures/repeated_measure.png" class="center" style="width:100.0%"></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"data_Yuan2006PMCBioinf.csv"</span>, package <span class="op">=</span> <span class="st">"rtpcr"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">data</span></span>
<span></span>
<span><span class="co"># Anova analysis</span></span>
<span><span class="fu"><a href="reference/ANOVA_DDCt.html">ANOVA_DDCt</a></span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  mainFactor.column <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  numOfFactors <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  numberOfrefGenes <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  block <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># An example of a properly arranged dataset from a repeated-measures experiment.</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"data_repeated_measure_1.csv"</span>, package <span class="op">=</span> <span class="st">"rtpcr"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">data</span></span>
<span></span>
<span><span class="co"># time  id  E_Target    Ct_target   E_Ref      Ct_Ref</span></span>
<span><span class="co">#    1   1      2       18.92   2   32.77</span></span>
<span><span class="co">#    1   2      2       15.82   2   32.45</span></span>
<span><span class="co">#    1   3      2       19.84   2   31.62</span></span>
<span><span class="co">#    2   1      2       19.46   2   33.03</span></span>
<span><span class="co">#    2   2      2       17.56   2   33.24</span></span>
<span><span class="co">#    2   3      2       19.74   2   32.08</span></span>
<span><span class="co">#    3   1      2       15.73   2   32.95</span></span>
<span><span class="co">#    3   2      2       17.21   2   33.64</span></span>
<span><span class="co">#    3   3      2       18.09   2   33.40</span></span>
<span></span>
<span><span class="co"># Repeated measure analysis</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ANOVA_DDCt.html">ANOVA_DDCt</a></span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  numOfFactors <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  numberOfrefGenes <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  mainFactor.column <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  block <span class="op">=</span> <span class="cn">NULL</span>, model <span class="op">=</span> <span class="va">wDCt</span> <span class="op">~</span> <span class="va">time</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">id</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Paired t.test (equivalent to repeated measure analysis, but not always the same results, due to different calculation methods!)</span></span>
<span><span class="fu"><a href="reference/TTEST_DDCt.html">TTEST_DDCt</a></span><span class="op">(</span></span>
<span>  <span class="va">data</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,<span class="op">]</span>, </span>
<span>  numberOfrefGenes <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  paired <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Anova analysis</span></span>
<span><span class="va">data3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"data_2factorBlock3ref.csv"</span>, package <span class="op">=</span> <span class="st">"rtpcr"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ANOVA_DDCt.html">ANOVA_DDCt</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">data3</span>,</span>
<span>  mainFactor.column <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  numOfFactors <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  numberOfrefGenes <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  block <span class="op">=</span> <span class="st">"block"</span>,</span>
<span>  analyseAllTarget <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h1>
<div class="section level2">
<h2 id="data-output">Data output<a class="anchor" aria-label="anchor" href="#data-output"></a>
</h2>
<p>All the functions for relative expression analysis (including <code>TTEST_DDCt</code>, <code>WILCOX_DDCt</code>, <code><a href="reference/ANOVA_DDCt.html">ANOVA_DDCt()</a></code>, and <code><a href="reference/ANOVA_DCt.html">ANOVA_DCt()</a></code>) return the relative expression table which include fold change and corresponding statistics. The output of <code><a href="reference/ANOVA_DDCt.html">ANOVA_DDCt()</a></code>, and <code><a href="reference/ANOVA_DCt.html">ANOVA_DCt()</a></code> also include lm model, residuals, raw data and ANOVA table for each gene. These outputs can be obtained as follow:</p>
<table class="table">
<thead><tr class="header">
<th>Per_gene Output</th>
<th>Code</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>expression table</td>
<td><code>res$relativeExpression</code></td>
</tr>
<tr class="even">
<td>ANOVA table</td>
<td><code>res$perGene$gene_name$ANOVA_table</code></td>
</tr>
<tr class="odd">
<td>ANOVA lm</td>
<td><code>res$perGene$gene_name$lm</code></td>
</tr>
<tr class="even">
<td>ANOVA lm formula</td>
<td><code>res$perGene$gene_name$lm_formula</code></td>
</tr>
<tr class="odd">
<td>Residuals</td>
<td><code>resid(res$perGene$gene_name$lm)</code></td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Relative expression table for the specified column in the input data:</span></span>
<span><span class="va">data3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"data_2factorBlock3ref.csv"</span>, package <span class="op">=</span> <span class="st">"rtpcr"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ANOVA_DDCt.html">ANOVA_DDCt</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">data3</span>,</span>
<span>  mainFactor.column <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  numOfFactors <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  numberOfrefGenes <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  block <span class="op">=</span> <span class="st">"block"</span>,</span>
<span>  analyseAllTarget <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Relative Expression</span></span>
<span><span class="co">#   gene contrast     ddCt      RE   log2FC     LCL     UCL      se Lower.se.RE Upper.se.RE Lower.se.log2FC Upper.se.log2FC  pvalue sig</span></span>
<span><span class="co"># 1   PO       L1  0.00000 1.00000  0.00000 0.00000 0.00000 0.13940     0.90790     1.10144         0.00000         0.00000 1.00000    </span></span>
<span><span class="co"># 2   PO L2 vs L1 -0.94610 1.92666  0.94610 1.25860 2.94934 0.14499     1.74245     2.13036         0.85564         1.04613 0.00116  **</span></span>
<span><span class="co"># 3   PO L3 vs L1 -2.19198 4.56931  2.19198 3.08069 6.77724 0.29402     3.72685     5.60221         1.78783         2.68748 0.00000 ***</span></span>
<span><span class="co"># 4  NLM       L1  0.00000 1.00000  0.00000 0.00000 0.00000 0.91809     0.52921     1.88962         0.00000         0.00000 1.00000    </span></span>
<span><span class="co"># 5  NLM L2 vs L1  0.86568 0.54879 -0.86568 0.39830 0.75614 0.36616     0.42577     0.70734        -1.11579        -0.67163 0.00018 ***</span></span>
<span><span class="co"># 6  NLM L3 vs L1 -1.44341 2.71964  1.44341 1.94670 3.79946 0.17132     2.41511     3.06256         1.28179         1.62542 0.00000 ***</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># The L1 level was used as calibrator.</span></span>
<span><span class="co"># Note: Using default model for statistical analysis: wDCt ~ block + Concentration * Type </span></span>
<span></span>
<span></span>
<span><span class="va">ANOVA_table</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">perGene</span><span class="op">$</span><span class="va">PO</span><span class="op">$</span><span class="va">ANOVA_table</span></span>
<span><span class="va">ANOVA_table</span></span>
<span></span>
<span><span class="va">lm</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">perGene</span><span class="op">$</span><span class="va">PO</span><span class="op">$</span><span class="va">lm</span></span>
<span><span class="va">lm</span></span>
<span></span>
<span><span class="va">lm_formula</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">perGene</span><span class="op">$</span><span class="va">gene_name</span><span class="op">$</span><span class="va">lm_formula</span></span>
<span><span class="va">lm_formula</span></span>
<span></span>
<span><span class="va">residuals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/residuals.html" class="external-link">resid</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">perGene</span><span class="op">$</span><span class="va">gene_name</span><span class="op">$</span><span class="va">lm</span><span class="op">)</span></span>
<span><span class="va">residuals</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="plot-output">Plot output<a class="anchor" aria-label="anchor" href="#plot-output"></a>
</h2>
<p>A single function of <code>plotFactor</code> is used to produce barplots for one- to three-factor expression tables.</p>
</div>
<div class="section level2">
<h2 id="plot-output-example-1">Plot output: example 1<a class="anchor" aria-label="anchor" href="#plot-output-example-1"></a>
</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"data_3factor.csv"</span>, package <span class="op">=</span> <span class="st">"rtpcr"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#Perform analysis first</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ANOVA_DCt.html">ANOVA_DCt</a></span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  numOfFactors <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  numberOfrefGenes <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  block <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span>  </span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">relativeExpression</span></span>
<span><span class="va">df</span></span>
<span> <span class="co"># Generate three-factor bar plot</span></span>
<span><span class="fu"><a href="reference/plotFactor.html">plotFactor</a></span><span class="op">(</span></span>
<span>  <span class="va">df</span>,</span>
<span>  x_col <span class="op">=</span> <span class="st">"SA"</span>,       </span>
<span>  y_col <span class="op">=</span> <span class="st">"log2FC"</span>,       </span>
<span>  group_col <span class="op">=</span> <span class="st">"Type"</span>,   </span>
<span>  facet_col <span class="op">=</span> <span class="st">"Conc"</span>,    </span>
<span>  Lower.se_col <span class="op">=</span> <span class="st">"Lower.se.log2FC"</span>,</span>
<span>  Upper.se_col <span class="op">=</span> <span class="st">"Upper.se.log2FC"</span>,</span>
<span>  letters_col <span class="op">=</span> <span class="st">"sig"</span>,</span>
<span>  letters_d <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  col_width <span class="op">=</span> <span class="fl">0.7</span>, </span>
<span>  dodge_width <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  fill_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"palegreen3"</span>, <span class="st">"skyblue"</span><span class="op">)</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  base_size <span class="op">=</span> <span class="fl">14</span>, </span>
<span>  alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  legend_position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/Rplot02.png" class="center" height="400"></p>
</div>
</div>
<div class="section level1">
<h1 id="how-to-edit-ouptput-plots">How to edit ouptput plots?<a class="anchor" aria-label="anchor" href="#how-to-edit-ouptput-plots"></a>
</h1>
<p>the rtpcr <code>plotFactor</code> function create ggplot objects for one to three factor tables that can further be edited by adding new layers:</p>
<table class="table">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr class="header">
<th>Task</th>
<th>Example Code</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Change y-axis label</td>
<td><code>p + ylab("Relative expression ($\Delta\Delta Ct$ method)")</code></td>
</tr>
<tr class="even">
<td>Add a horizontal reference line</td>
<td><code>p + geom_hline(yintercept = 0, linetype = "dashed")</code></td>
</tr>
<tr class="odd">
<td>Change y-axis limits</td>
<td><code>p + scale_y_continuous(expand = expansion(mult = c(0, 0.1)))</code></td>
</tr>
<tr class="even">
<td>Relabel x-axis</td>
<td><code>p + scale_x_discrete(labels = c("A" = "Control", "B" = "Treatment"))</code></td>
</tr>
<tr class="odd">
<td>Change fill colors</td>
<td><code>p + scale_fill_brewer(palette = "Set2")</code></td>
</tr>
</tbody>
</table>
<div class="section level3">
<h3 id="plot-output-example-2">Plot output: example 2<a class="anchor" aria-label="anchor" href="#plot-output-example-2"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"data_2factorBlock.csv"</span>, package <span class="op">=</span> <span class="st">"rtpcr"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ANOVA_DCt.html">ANOVA_DCt</a></span><span class="op">(</span><span class="va">data</span>, </span>
<span>      numOfFactors <span class="op">=</span> <span class="fl">2</span>,</span>
<span>      block <span class="op">=</span> <span class="st">"block"</span>,</span>
<span>      numberOfrefGenes <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">relativeExpression</span></span>
<span></span>
<span><span class="fu"><a href="reference/plotFactor.html">plotFactor</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">df</span>,</span>
<span>  x_col <span class="op">=</span> <span class="st">"factor2"</span>,</span>
<span>  y_col <span class="op">=</span> <span class="st">"RE"</span>,</span>
<span>  group_col <span class="op">=</span> <span class="st">"factor1"</span>,</span>
<span>  Lower.se_col <span class="op">=</span> <span class="st">"Lower.se.RE"</span>,</span>
<span>  Upper.se_col <span class="op">=</span> <span class="st">"Upper.se.RE"</span>,</span>
<span>  letters_col <span class="op">=</span> <span class="st">"sig"</span>,</span>
<span>  letters_d <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  fill_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"aquamarine4"</span>, <span class="st">"gold2"</span><span class="op">)</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  col_width <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  dodge_width <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  base_size <span class="op">=</span> <span class="fl">16</span>, </span>
<span>  legend_position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.8</span>, <span class="fl">0.8</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/Rplot01.png" class="center" height="300"></p>
</div>
<div class="section level3">
<h3 id="plot-output-example-3">Plot output: example 3<a class="anchor" aria-label="anchor" href="#plot-output-example-3"></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Heffer et al., 2020, PlosOne</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"data_Heffer2020PlosOne.csv"</span>, package <span class="op">=</span> <span class="st">"rtpcr"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ANOVA_DDCt.html">ANOVA_DDCt</a></span><span class="op">(</span></span>
<span>  <span class="va">df</span>,</span>
<span>  numOfFactors <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  mainFactor.column <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  numberOfrefGenes <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  block <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">relativeExpression</span></span>
<span></span>
<span><span class="co"># Selecting only the first words in 'contrast' column to be used as the x-axis labels.</span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">contrast</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">" .*"</span>, <span class="st">""</span>, <span class="va">data</span><span class="op">$</span><span class="va">contrast</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/plotFactor.html">plotFactor</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">data</span>,</span>
<span>  x_col <span class="op">=</span> <span class="st">"contrast"</span>,</span>
<span>  y_col <span class="op">=</span> <span class="st">"RE"</span>,</span>
<span>  group_col <span class="op">=</span> <span class="st">"contrast"</span>,</span>
<span>  facet_col <span class="op">=</span> <span class="st">"gene"</span>,</span>
<span>  Lower.se_col <span class="op">=</span> <span class="st">"Lower.se.RE"</span>,</span>
<span>  Upper.se_col <span class="op">=</span> <span class="st">"Upper.se.RE"</span>,</span>
<span>  letters_col <span class="op">=</span> <span class="st">"sig"</span>,</span>
<span>  letters_d <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  fill_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palette.html" class="external-link">palette.colors</a></span><span class="op">(</span><span class="fl">4</span>, recycle <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  col_width <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  dodge_width <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  base_size <span class="op">=</span> <span class="fl">16</span>, </span>
<span>  legend_position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/Rplot03.png" class="center" height="700"></p>
</div>
<div class="section level3">
<h3 id="plot-output-example-4">Plot output: example 4<a class="anchor" aria-label="anchor" href="#plot-output-example-4"></a>
</h3>
<p>The function <code><a href="reference/plotSingleGene.html">plotSingleGene()</a></code> creates a bar plot of relative gene expression (fold change) values from single gene analysis showing all pairwise significances.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">ANOVA_DDCt</span>(</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>  data_Heffer2020PlosOne,</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>  <span class="at">numOfFactors =</span> <span class="dv">1</span>,</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>  <span class="at">mainFactor.column =</span> <span class="dv">1</span>,</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>  <span class="at">numberOfrefGenes =</span> <span class="dv">1</span>,</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>  <span class="at">block =</span> <span class="cn">NULL</span>,</span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>  <span class="at">analyseAllTarget =</span> <span class="st">"Tnfa"</span>) </span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a><span class="fu">plotSingleGene</span>(res, <span class="at">fill =</span> <span class="st">"cyan4"</span>, <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">base_size =</span> <span class="dv">12</span>)</span></code></pre></div>
<p><img src="reference/figures/signif.png" class="center" height="400"></p>
</div>
</div>
<div class="section level1">
<h1 id="post-hoc-analysis">Post-hoc analysis<a class="anchor" aria-label="anchor" href="#post-hoc-analysis"></a>
</h1>
<p>Although all the expression analysis functions perform statistical comparisons for the levels of the analysed factor, further Post-hoc analysis is still possible. The <code>Means_DDCt</code> function performs post-hoc comparisons using a fitted model object produced by <code>ANOVA_DCt</code> and <code>ANOVA_DDCt</code>. It applies pairwise statistical comparisons of relative expression (RE) values for user-specified effects via the <code>specs</code> argument. Supported effects include simple effects, interactions, and slicing, provided the underlying model is an ANOVA. For ANCOVA models returned by this package, the <code>Means_DDCt</code> output is limited to simple effects only.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ANOVA_DDCt.html">ANOVA_DDCt</a></span><span class="op">(</span></span>
<span>  <span class="va">data_3factor</span>,</span>
<span>  numOfFactors <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  numberOfrefGenes <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  mainFactor.column <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  block <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="va">model</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">perGene</span><span class="op">$</span><span class="va">E_PO</span><span class="op">$</span><span class="va">lm</span></span>
<span><span class="co"># Relative expression values for Concentration main effect</span></span>
<span><span class="fu"><a href="reference/Means_DDCt.html">Means_DDCt</a></span><span class="op">(</span><span class="va">model</span>, specs <span class="op">=</span> <span class="st">"Conc"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># contrast        RE        SE df       LCL       UCL p.value sig</span></span>
<span><span class="co"># L vs H   0.1703610 0.2208988 24 0.1242014 0.2336757 &lt;0.0001 ***</span></span>
<span><span class="co"># M vs H   0.2227247 0.2208988 24 0.1623772 0.3055004 &lt;0.0001 ***</span></span>
<span><span class="co"># M vs L   1.3073692 0.2208988 24 0.9531359 1.7932535  0.0928 .  </span></span>
<span><span class="co">#</span></span>
<span><span class="co">#Results are averaged over the levels of: Type, SA </span></span>
<span><span class="co">#Confidence level used: 0.95 </span></span>
<span></span>
<span><span class="co"># Relative expression values for Concentration sliced by Type</span></span>
<span><span class="fu"><a href="reference/Means_DDCt.html">Means_DDCt</a></span><span class="op">(</span><span class="va">model</span>, specs <span class="op">=</span> <span class="st">"Conc | Type"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Type = R:</span></span>
<span><span class="co">#  contrast       RE        SE df       LCL      UCL p.value sig</span></span>
<span><span class="co">#  L vs H   0.103187 0.3123981 24 0.0659984 0.161331 &lt;0.0001 ***</span></span>
<span><span class="co">#  M vs H   0.339151 0.3123981 24 0.2169210 0.530255 &lt;0.0001 ***</span></span>
<span><span class="co">#  M vs L   3.286761 0.3123981 24 2.1022126 5.138776 &lt;0.0001 ***</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># Type = S:</span></span>
<span><span class="co">#  contrast       RE        SE df       LCL      UCL p.value sig</span></span>
<span><span class="co">#  L vs H   0.281265 0.3123981 24 0.1798969 0.439751 &lt;0.0001 ***</span></span>
<span><span class="co">#  M vs H   0.146266 0.3123981 24 0.0935518 0.228684 &lt;0.0001 ***</span></span>
<span><span class="co">#  M vs L   0.520030 0.3123981 24 0.3326112 0.813055  0.0059 ** </span></span>
<span><span class="co">#</span></span>
<span><span class="co"># Results are averaged over the levels of: SA </span></span>
<span><span class="co"># Confidence level used: 0.95 </span></span>
<span></span>
<span><span class="co"># Relative expression values for Concentration sliced by Type and SA</span></span>
<span><span class="fu"><a href="reference/Means_DDCt.html">Means_DDCt</a></span><span class="op">(</span><span class="va">model</span>, specs <span class="op">=</span> <span class="st">"Conc | Type * SA"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level1">
<h1 id="checking-normality-of-residuals">Checking normality of residuals<a class="anchor" aria-label="anchor" href="#checking-normality-of-residuals"></a>
</h1>
<p>If the residuals from a <code>t.test</code> or an <code>lm</code> object are not normally distributed, the significance results might be violated. In such cases, non-parametric tests can be used. For example, the Mann–Whitney test - also known as the Wilcoxon rank-sum test, (implemented via <code><a href="reference/WILCOX_DDCt.html">WILCOX_DDCt()</a></code> in the rtpcr package), is an alternative to t.test, and <code><a href="https://rdrr.io/r/stats/kruskal.test.html" class="external-link">kruskal.test()</a></code> is an alternative to one-way analysis of variance. These tests assess differences between population medians using independent samples. However, the <code>t.test</code> function (also the <code>TTEST_DDCt</code> function described above) includes the <code>var.equal</code> argument. When set to <code>FALSE</code>, performs <code>t.test</code> under the unequal variances hypothesis. Residuals from <code>ANOVA_DCt</code> and <code>ANOVA_DDCt</code> functions objects can be extracted from <code>lm</code>and plotted as follow:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"data_repeated_measure_1.csv"</span>, package <span class="op">=</span> <span class="st">"rtpcr"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">res3</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ANOVA_DDCt.html">ANOVA_DDCt</a></span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  numOfFactors <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  numberOfrefGenes <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  mainFactor.column <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  block <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  model <span class="op">=</span> <span class="va">wDCt</span> <span class="op">~</span> <span class="va">time</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">id</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">residuals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/residuals.html" class="external-link">resid</a></span><span class="op">(</span><span class="va">res3</span><span class="op">$</span><span class="va">perGene</span><span class="op">$</span><span class="va">Target</span><span class="op">$</span><span class="va">lm</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/shapiro.test.html" class="external-link">shapiro.test</a></span><span class="op">(</span><span class="va">residuals</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">residuals</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/qqnorm.html" class="external-link">qqnorm</a></span><span class="op">(</span><span class="va">residuals</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/qqnorm.html" class="external-link">qqline</a></span><span class="op">(</span><span class="va">residuals</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level1">
<h1 id="mean-of-technical-replicates">Mean of technical replicates<a class="anchor" aria-label="anchor" href="#mean-of-technical-replicates"></a>
</h1>
<p>Calculating the mean of technical replicates and generating an output table suitable for subsequent ANOVA analysis can be accomplished using the <code>meanTech</code> function. The input dataset must follow the column structure illustrated in the example data below. Columns used for grouping should be explicitly specified via the <code>groups</code> argument of the <code>meanTech</code> function.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example input data frame with technical replicates</span></span>
<span><span class="va">data1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"data_withTechRep.csv"</span>, package <span class="op">=</span> <span class="st">"rtpcr"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate mean of technical replicates using first four columns as groups</span></span>
<span><span class="fu"><a href="reference/meanTech.html">meanTech</a></span><span class="op">(</span><span class="va">data1</span>,</span>
<span>         groups <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,</span>
<span>         numOfFactors <span class="op">=</span> <span class="fl">1</span>,</span>
<span>         block <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/techrep.png" class="center" height="380"></p>
</div>
<div class="section level1">
<h1 id="contact">Contact<a class="anchor" aria-label="anchor" href="#contact"></a>
</h1>
<p>Email: gh.mirzaghaderi at uok.ac.ir</p>
</div>
<div class="section level1">
<h1 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h1>
<div class="sourceCode" id="cb13"><pre class="sourceCode md"><code class="sourceCode markdown"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>citation("rtpcr")</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>To cite the package ‘rtpcr’ in publications, please use:</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>  Ghader Mirzaghaderi (2025). rtpcr: a package for statistical analysis and graphical</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>  presentation of qPCR data in R. PeerJ 13:e20185. https://doi.org/10.7717/peerj.20185</span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a>A BibTeX entry for LaTeX users is</span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a></span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a>  @Article{,</span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a>    title = {rtpcr: A package for statistical analysis and graphical presentation of qPCR data in R},</span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a>    author = {Ghader Mirzaghaderi},</span>
<span id="cb13-13"><a href="#cb13-13" tabindex="-1"></a>    journal = {PeerJ},</span>
<span id="cb13-14"><a href="#cb13-14" tabindex="-1"></a>    volume = {13},</span>
<span id="cb13-15"><a href="#cb13-15" tabindex="-1"></a>    pages = {e20185},</span>
<span id="cb13-16"><a href="#cb13-16" tabindex="-1"></a>    year = {2025},</span>
<span id="cb13-17"><a href="#cb13-17" tabindex="-1"></a>    doi = {10.7717/peerj.20185},</span>
<span id="cb13-18"><a href="#cb13-18" tabindex="-1"></a>  }</span></code></pre></div>
</div>
<div class="section level1">
<h1 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h1>
<p>Livak, Kenneth J, and Thomas D Schmittgen. 2001. Analysis of Relative Gene Expression Data Using Real-Time Quantitative PCR and the Double Delta CT Method. Methods 25 (4). <a href="https://doi.org/10.1006/meth.2001.1262" class="external-link">doi.org/10.1006/meth.2001.1262</a>.</p>
<p>Ganger, MT, Dietz GD, Ewing SJ. 2017. A common base method for analysis of qPCR data and the application of simple blocking in qPCR experiments. BMC bioinformatics 18, 1-11. <a href="https://doi.org/10.1186/s12859-017-1949-5" class="external-link">doi.org/10.1186/s12859-017-1949-5</a>.</p>
<p>Mirzaghaderi G. 2025. rtpcr: a package for statistical analysis and graphical presentation of qPCR data in R. PeerJ 13, e20185. <a href="https://doi.org/10.7717/peerj.20185" class="external-link">doi.org/10.7717/peerj.20185</a>.</p>
<p>Pfaffl MW, Horgan GW, Dempfle L. 2002. Relative expression software tool (REST©) for group-wise comparison and statistical analysis of relative expression results in real-time PCR. Nucleic acids research 30, e36-e36. <a href="https://doi.org/10.1093/nar/30.9.e36" class="external-link">doi.org/10.1093/nar/30.9.e36</a>.</p>
<p>Taylor SC, Nadeau K, Abbasi M, Lachance C, Nguyen M, Fenrich, J. 2019. The ultimate qPCR experiment: producing publication quality, reproducible data the first time. Trends in Biotechnology, 37(7), 761-774. <a href="https://doi.org/10.1016/j.tibtech.2018.12.002" class="external-link">doi.org/10.1016/j.tibtech.2018.12.002</a>.</p>
<p>Yuan, JS, Ann Reed, Feng Chen, and Neal Stewart. 2006. Statistical Analysis of Real-Time PCR Data. BMC Bioinformatics 7 (85). <a href="https://doi.org/10.1186/1471-2105-7-85" class="external-link">doi.org/10.1186/1471-2105-7-85</a>.</p>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=rtpcr" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/mirzaghaderi/rtpcr/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing rtpcr</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Ghader Mirzaghaderi <br><small class="roles"> Author, maintainer, copyright holder </small>   </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://cran.rstudio.com/web/packages/rtpcr/index.html" class="external-link"><img src="https://cranlogs.r-pkg.org/badges/grand-total/rtpcr"></a></li>
<li><a href="https://github.com/mirzaghaderi/rtpcr" class="external-link"><img src="https://img.shields.io/github/r-package/v/mirzaghaderi/rtpcr" alt="R package version"></a></li>
<li><a href="https://CRAN.R-project.org/package=rtpcr" class="external-link"><img src="https://www.r-pkg.org/badges/version/rtpcr" alt="CRAN version"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ghader Mirzaghaderi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
