<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>qPCR Data Analysis ‚Ä¢ rtpcr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="qPCR Data Analysis">
<meta property="og:description" content="Tools for analysis of RT-qPCR gene expression data using ŒîCt and ŒîŒîCt methods, including t-tests, ANOVA, ANCOVA, repeated-measures models, and publication-ready visualizations. The package implements a general calculation framework described by Ganger et al. (2017) &lt;doi:10.1186/s12859-017-1949-5&gt; and Taylor et al. (2019) &lt;doi:10.1016/j.tibtech.2018.12.002&gt;, covering both the Livak and Pfaffl methods.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">rtpcr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mirzaghaderi/rtpcr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">


<div class="section level1">
<div class="page-header"><h1 id="multi-target-multi-reference-qpcr-data-analysis-using-rtpcr-package">Multi-target multi-reference qPCR data analysis using <strong><code>rtpcr</code></strong> package<a class="anchor" aria-label="anchor" href="#multi-target-multi-reference-qpcr-data-analysis-using-rtpcr-package"></a>
</h1></div>
<ul>
<li><a href="#overview">Overview</a></li>
<li><a href="#improvement-in-the-current-github-version-rtpcr_v211">Improvement in the current GitHub version <code>rtpcr_v2.1.1</code></a></li>
<li><a href="#functions">Functions</a></li>
<li>
<a href="#quick-start">Quick start</a>
<ul>
<li><a href="#installing-and-loading">Installing and loading</a></li>
</ul>
</li>
<li><a href="#input-data-structure">Input data structure</a></li>
<li><a href="#handling-missing-ct-values">Handling missing Ct values</a></li>
<li>
<a href="#data-analysis">Data analysis</a>
<ul>
<li><a href="#amplification-efficiency">Amplification efficiency</a></li>
<li><a href="#relative-expression">Relative expression</a></li>
</ul>
</li>
<li>
<a href="#output">Output</a>
<ul>
<li><a href="#data-output">Data output</a></li>
<li><a href="#plot-output">Plot output</a></li>
</ul>
</li>
<li><a href="#post-hoc-analysis">Post-hoc analysis</a></li>
<li><a href="#checking-normality-of-residuals">Checking normality of residuals</a></li>
<li><a href="#mean-of-technical-replicates">Mean of technical replicates</a></li>
<li><a href="#contact">Contact</a></li>
<li><a href="#citation">Citation</a></li>
<li><a href="#references">References</a></li>
</ul>
</div>
<div class="section level1">
<h1 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h1>
<p>Tools for analysis of RT-qPCR gene expression data using ŒîCt and ŒîŒîCt methods, including t-tests, ANOVA, ANCOVA, repeated-measures models, and publication-ready visualizations. The package implements a <a href="inst/method.html">general calculation method</a> described by Ganger et al.¬†(2017) and Taylor et al.¬†(2019), covering both the Livak and Pfaffl methods.</p>
<div class="section level2">
<h2 id="improvement-in-the-current-github-version-rtpcr_v211">Improvement in the current GitHub version <code>rtpcr_v2.1.1</code>:<a class="anchor" aria-label="anchor" href="#improvement-in-the-current-github-version-rtpcr_v211"></a>
</h2>
<p>‚úîÔ∏è <strong>A unified data structure as input for all functions.</strong></p>
<p>üî¥ <span style="color: green;"><strong>No restriction on the number of target or reference genes.</strong></span></p>
<p>üî¥ <span style="color: green;"><strong>Supports analysis of all genes or selected subsets.</strong></span></p>
<p>üî¥ <span style="color: green;"><strong>Publication-ready graphical outputs.</strong></span></p>
</div>
</div>
<div class="section level1">
<h1 id="functions">Functions<a class="anchor" aria-label="anchor" href="#functions"></a>
</h1>
<p>The <code>rtpcr</code> package takes a table of efficiency (E) the Ct values of genes and performs different analyses using the following functions.</p>
<table class="table">
<colgroup>
<col width="25%">
<col width="74%">
</colgroup>
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>ANOVA_DCt</code></td>
<td>ŒîCt ANOVA analysis</td>
</tr>
<tr class="even">
<td><code>ANOVA_DDCt</code></td>
<td>ŒîŒîCt ANOVA analysis</td>
</tr>
<tr class="odd">
<td><code>REPEATED_DDCt</code></td>
<td>ŒîŒîCt ANOVA analysis for repeated-measures data</td>
</tr>
<tr class="even">
<td><code>TTEST_DDCt</code></td>
<td>ŒîŒîCt method <em>t</em>-test analysis</td>
</tr>
<tr class="odd">
<td><code>plotFactor</code></td>
<td>Bar plot of gene expression for one-, two- or three-factor experiments</td>
</tr>
<tr class="even">
<td><code>Means_DDCt</code></td>
<td>Pairwise comparison of RE values for any user-specified effect</td>
</tr>
<tr class="odd">
<td><code>efficiency</code></td>
<td>Amplification efficiency statistics and standard curves</td>
</tr>
<tr class="even">
<td><code>meanTech</code></td>
<td>Calculate mean of technical replicates</td>
</tr>
<tr class="odd">
<td><code>multiplot</code></td>
<td>Combine multiple ggplot objects into a single layout</td>
</tr>
</tbody>
</table>
</div>
<div class="section level1">
<h1 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h1>
<div class="section level3">
<h3 id="installing-and-loading">Installing and loading<a class="anchor" aria-label="anchor" href="#installing-and-loading"></a>
</h3>
<p>The current version of the <code>rtpcr</code> package can be installed from GitHub by running the following code in R:</p>
<p>```{r devtools::install_github(‚Äúmirzaghaderi/rtpcr‚Äù, build_vignettes = FALSE)</p>
</div>
</div>
<div class="section level1">
<h1 id="loading-the-package">Loading the package<a class="anchor" aria-label="anchor" href="#loading-the-package"></a>
</h1>
<p>library(rtpcr)</p>
<pre><code>
Or from CRAN which may install the previous version.
```{r
# Installing from CRAN
install.packages("rtpcr")</code></pre>
</div>
<div class="section level1">
<h1 id="input-data-structure">Input data structure<a class="anchor" aria-label="anchor" href="#input-data-structure"></a>
</h1>
<p>Input data structure is important and should be in wide format. For relative expression analysis (using <code>TTEST_DDCt</code>, <code>ANOVA_DCt</code>, <code>ANOVA_DDCt</code> and <code>REPEATED_DDCt</code> functions), the required column structure of the input data is:</p>
<ol style="list-style-type: decimal">
<li>Experimental condition columns (up to 3 factors, and one block if available)</li>
<li>Replicates information (biological replicates or subjects; see <a href="#note-1">NOTE 1</a>)<br>
</li>
<li>Target genes efficiency and Ct values (a pair column for each gene)</li>
<li>Reference genes efficiency and Ct values (a pair column for each gene)</li>
</ol>
<p>The package supports <strong>one or more target or reference gene(s)</strong>, supplied as efficiency‚ÄìCt column pairs. Reference gene columns must always appear last. A sample input data is presented below.</p>
<p align="center">
<img src="inst/dataStructure1.png" width="100%"></p>
<details><div class="float">
<img src="figures/dataStructure1.png" alt="Workflow of the analysis"><div class="figcaption">Workflow of the analysis</div>
</div>
</details><div class="section level4">
<h4 id="note-1">NOTE 1<a class="anchor" aria-label="anchor" href="#note-1"></a>
</h4>
<p>For <code>TTEST_DDCt</code>, <code>ANOVA_DCt</code>, and <code>ANOVA_DDCt</code>, each row is from a separate and uniq biological replicate. For example, a dataframe with 12 rows has come from an experiment with 12 individuals. The <code>REPEATED_DDCt</code> function is intended for experiments with repeated observations (e.g.¬†time-course data). For <code>REPEATED_DDCt</code>, the Replicate column contains identifiers for each individual (id or subject). For example, all rows with a <code>1</code> at Rep column correspond to a single individual, all rows with a <code>2</code> correspond to another individual, and so on, which have been sampled at specific time points.</p>
</div>
</div>
<div class="section level1">
<h1 id="handling-missing-ct-values">Handling missing Ct values<a class="anchor" aria-label="anchor" href="#handling-missing-ct-values"></a>
</h1>
<p>The <code>rtpcr</code> package does not automatically handle missing Ct values. However, NA can be used to represent missing Ct measurements in the input data. In such cases, NA is returned for the corresponding individual ŒîCt and passed along to downstream statistical analyses. If more sophisticated handling of missing Ct values (fixed value of 40 for target genes, mean Ct alues of other replicates, or imputation) is desired, qPCR imputation tools can be used in advance of analysis with rtpcr.</p>
</div>
<div class="section level1">
<h1 id="data-analysis">Data analysis<a class="anchor" aria-label="anchor" href="#data-analysis"></a>
</h1>
<div class="section level3">
<h3 id="amplification-efficiency">Amplification efficiency<a class="anchor" aria-label="anchor" href="#amplification-efficiency"></a>
</h3>
<p>The <code>efficiency</code> function calculates the amplification efficiency (E), slope, and R¬≤ statistics for genes, and performs pairwise comparisons of slopes. It takes a data frame in which the first column contains the dilution ratios, followed by the Ct value columns for each gene.</p>
<p>```{r eval= F} # Applying the efficiency function data &lt;- read.csv(system.file(‚Äúextdata‚Äù, ‚Äúdata_efficiency.csv‚Äù, package = ‚Äúrtpcr‚Äù)) data dilutions Gene1 Gene2 Gene3 1.00 25.58 24.25 22.61 1.00 25.54 24.13 22.68 1.00 25.50 24.04 22.63 0.50 26.71 25.56 23.67 0.50 26.73 25.43 23.65 0.50 26.87 26.01 23.70 0.20 28.17 27.37 25.11 0.20 28.07 26.94 25.12 0.20 28.11 27.14 25.11 0.10 29.20 28.05 26.17 0.10 29.49 28.89 26.15 0.10 29.07 28.32 26.15 0.05 30.17 29.50 27.12 0.05 30.14 29.93 27.14 0.05 30.12 29.71 27.16 0.02 31.35 30.69 28.52 0.02 31.35 30.54 28.57 0.02 31.35 30.04 28.53 0.01 32.55 31.12 29.49 0.01 32.45 31.29 29.48 0.01 32.28 31.15 29.26</p>
</div>
</div>
<div class="section level1">
<h1 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h1>
<p>efficiency(data)</p>
<p>$Efficiency Gene Slope R2 E 1 Gene1 -3.388094 0.9965504 1.973110 2 Gene2 -3.528125 0.9713914 1.920599 3 Gene3 -3.414551 0.9990278 1.962747</p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>S</mi><mi>l</mi><mi>o</mi><mi>p</mi><msub><mi>e</mi><mi>c</mi></msub><mi>o</mi><mi>m</mi><mi>p</mi><mi>a</mi><mi>r</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">Slope_compare</annotation></semantics></math>contrasts contrast estimate SE df t.ratio p.value C2H2.26 - C2H2.01 0.1400 0.121 57 1.157 0.4837 C2H2.26 - GAPDH 0.0265 0.121 57 0.219 0.9740 C2H2.01 - GAPDH -0.1136 0.121 57 -0.938 0.6186</p>
<pre><code>
### Relative expression
`TTEST_DDCt`, `ANOVA_DDCt`, `REPEATED_DDCt`, and `ANOVA_DCt`  functions perform relative expression analysis using ŒîŒîCt or ŒîCt methods. Below is an example of expression analysis using ŒîŒîCt method.
```{r eval= F}
# An example of a properly arranged dataset from a repeated-measures experiment.
data &lt;- read.csv(system.file("extdata", "data_repeated_measure_1.csv", package = "rtpcr"))
data

time    id  E_Target    Ct_target   E_Ref      Ct_Ref
   1     1      2       18.92   2   32.77
   1     2      2       15.82   2   32.45
   1     3      2       19.84   2   31.62
   2     1      2       19.46   2   33.03
   2     2      2       17.56   2   33.24
   2     3      2       19.74   2   32.08
   3     1      2       15.73   2   32.95
   3     2      2       17.21   2   33.64
   3     3      2       18.09   2   33.40

# Repeated measure analysis
res &lt;- REPEATED_DDCt(
  data,
  numOfFactors = 1,
  numberOfrefGenes = 1,
  repeatedFactor = "time",
  calibratorLevel = "1",
  block = NULL)


# Anova analysis
ANOVA_DDCt(
  data,
  mainFactor.column = 1,
  numOfFactors = 1,
  numberOfrefGenes = 1,
  block = NULL)


# Paired t.test (equivalent to repeated measure analysis, but not always the same results, due to different calculation methods!)
TTEST_DDCt(
  data[1:6,],
  numberOfrefGenes = 1,
  paired = T)


# Anova analysis
data &lt;- read.csv(system.file("extdata", "data_2factorBlock3ref.csv", package = "rtpcr"))
res &lt;- ANOVA_DDCt(
  x = data,
  mainFactor.column = 1,
  numOfFactors = 2,
  numberOfrefGenes = 1,
  block = "block",
  analyseAllTarget = TRUE)</code></pre>
</div>
<div class="section level1">
<h1 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h1>
<div class="section level2">
<h2 id="data-output">Data output<a class="anchor" aria-label="anchor" href="#data-output"></a>
</h2>
<p>A lot of outputs including relative expression table, lm models, residuals, raw data and ANOVA table for each gene can be accessed. The expression table of all genes is returned by <code>res$combinedFoldChange</code>. Other outpus for each gene can be obtained as follow:</p>
<table class="table">
<colgroup>
<col width="26%">
<col width="73%">
</colgroup>
<thead><tr class="header">
<th>Per_gene Output</th>
<th>Code</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>expression table</td>
<td><code>res$combinedFoldChange</code></td>
</tr>
<tr class="even">
<td>ANOVA table</td>
<td><code>res$perGene$gene_name$ANOVA_table</code></td>
</tr>
<tr class="odd">
<td>ANOVA lm</td>
<td><code>res$perGene$gene_name$lm_ANOVA</code></td>
</tr>
<tr class="even">
<td>ANCOVA lm</td>
<td><code>res$perGene$gene_name$lm_ANCOVA</code></td>
</tr>
<tr class="odd">
<td>Residuals</td>
<td><code>resid(res$perGene$gene_name$lm_ANOVA)</code></td>
</tr>
</tbody>
</table>
<p><code>{r eval= F} # Relative expression table for the specified column in the input data: df &lt;- res$combinedFoldChange df Relative Expression gene   contrast       RE  log2FC pvalue sig    LCL     UCL     se Lower.se.RE Upper.se.RE Lower.se.log2FC Upper.se.log2FC PO            R   1.0000  0.0000 1.0000     0.0000  0.0000 0.5506      0.6828      1.4647          0.0000          0.0000 PO       S vs R  11.6130  3.5377 0.0001 *** 4.4233 30.4888 0.2286      9.9115     13.6066          3.0193          4.1450 GAPDH         R   1.0000  0.0000 1.0000     0.0000  0.0000 0.4815      0.7162      1.3962          0.0000          0.0000 GAPDH    S vs R   6.6852  2.7410 0.0001 *** 3.0687 14.5641 0.3820      5.1301      8.7118          2.1034          3.5719 ref2          R   1.0000  0.0000 1.0000     0.0000  0.0000 0.6928      0.6186      1.6164          0.0000          0.0000 ref2     S vs R   0.9372 -0.0936 0.9005     0.3145  2.7929 0.2414      0.7927      1.1079         -0.1107         -0.0792</code></p>
</div>
<div class="section level2">
<h2 id="plot-output">Plot output<a class="anchor" aria-label="anchor" href="#plot-output"></a>
</h2>
<p>A single function of <code>plotFactor</code> is used to produce barplots for one- to three-factor expression tables.</p>
<div class="section level3">
<h3 id="plot-output-example-1">Plot output: example 1<a class="anchor" aria-label="anchor" href="#plot-output-example-1"></a>
</h3>
<p>```{r eval= F, warning = F, fig.height = 7, fig.width = 12.5, fig.align = ‚Äòcenter‚Äô, warning = F} data &lt;- read.csv(system.file(‚Äúextdata‚Äù, ‚Äúdata_3factor.csv‚Äù, package = ‚Äúrtpcr‚Äù)) #Perform analysis first res &lt;- ANOVA_DCt( data, numOfFactors = 3, numberOfrefGenes = 1, block = NULL)</p>
<p>df &lt;- res$combinedResults df # Generate three-factor bar plot p &lt;- plotFactor( df, x_col = ‚ÄúSA‚Äù,<br>
y_col = ‚Äúlog2FC‚Äù,<br>
group_col = ‚ÄúType‚Äù,<br>
facet_col = ‚ÄúConc‚Äù,<br>
Lower.se_col = ‚ÄúLower.se.log2FC‚Äù, Upper.se_col = ‚ÄúUpper.se.log2FC‚Äù, letters_col = ‚Äúsig‚Äù, letters_d = 0.3, col_width = 0.7, dodge_width = 0.7, fill_colors = c(‚Äúpalegreen3‚Äù, ‚Äúskyblue‚Äù), color = ‚Äúblack‚Äù, base_size = 14, alpha = 1, legend_position = c(0.1, 0.2))</p>
<p>library(ggplot2) p + theme( panel.border = element_rect(color = ‚Äúblack‚Äù, linewidth = 0.5))</p>
<pre><code>
&lt;p align="center"&gt;
&lt;img src="inst/Rplot02.png" width="100%"&gt;
&lt;/p&gt;

&lt;details&gt;
![Workflow of the analysis](figures/Rplot02.png)
&lt;/details&gt;

# How to edit ouptput plots?
the `rtpcr` plotFactor function create ggplot objects for one to three factor table that can furtherbe edited by adding new layers:

| Task | Example Code |
|------|--------------|
| **Change y-axis label** | `p + ylab("Relative expression (ŒîŒîCt method)")` |
| **Add a horizontal reference line** | `p + geom_hline(yintercept = 0, linetype = "dashed")` |
| **Change y-axis limits** | `p + scale_y_continuous(expand = expansion(mult = c(0, 0.1)))` |
| **Relabel x-axis** | `p + scale_x_discrete(labels = c("A" = "Control", "B" = "Treatment"))` |
| **Change fill colors** | `p + scale_fill_brewer(palette = "Set2")` |


### Plot output: example 2
```{r eval= F, fig.height = 7, fig.width = 12.5, fig.align = 'center', warning = F}
data &lt;- read.csv(system.file("extdata", "data_2factorBlock.csv", package = "rtpcr"))
res &lt;- ANOVA_DCt(data,
      numOfFactors = 2,
      block = "block",
      numberOfrefGenes = 1)

df &lt;- res$combinedResults

p1 &lt;- plotFactor(
  data = df,
  x_col = "factor2",
  y_col = "RE",
  group_col = "factor1",
  Lower.se_col = "Lower.se.RE",
  Upper.se_col = "Upper.se.RE",
  letters_col = "sig",
  letters_d = 0.2,
  fill_colors = c("aquamarine4", "gold2"),
  color = "black",
  alpha = 1,
  col_width = 0.7,
  dodge_width = 0.7,
  base_size = 16,
  legend_position = c(0.2, 0.8))

library(ggplot2)
p1 +
  theme(axis.text.y = element_text(size = 14, color = "black")) +
  theme(axis.text.x = element_text(size = 14, color = "black", angle = 45, hjust = 1)) +
  theme(legend.text = element_text(colour = "black", size = 14),
        legend.background = element_rect(fill = "transparent")) +
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)))</code></pre>
<p align="center">
<img src="inst/Rplot01.png" width="100%"></p>
<details><div class="float">
<img src="figures/Rplot01.png" alt="Workflow of the analysis"><div class="figcaption">Workflow of the analysis</div>
</div>
</details>
</div>
<div class="section level3">
<h3 id="plot-output-example-3">Plot output: example 3<a class="anchor" aria-label="anchor" href="#plot-output-example-3"></a>
</h3>
<p>```{r eval= F, warning = F} # Heffer et al., 2020, PlosOne library(dplyr) df &lt;- read.csv(system.file(‚Äúextdata‚Äù, ‚ÄúHeffer2020PlosOne.csv‚Äù, package = ‚Äúrtpcr‚Äù))</p>
<p>res &lt;- ANOVA_DDCt( df, numOfFactors = 1, mainFactor.column = 1, numberOfrefGenes = 1, block = NULL)</p>
<p>data &lt;- res<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>c</mi><mi>o</mi><mi>m</mi><mi>b</mi><mi>i</mi><mi>n</mi><mi>e</mi><mi>d</mi><mi>F</mi><mi>o</mi><mi>l</mi><mi>d</mi><mi>C</mi><mi>h</mi><mi>a</mi><mi>n</mi><mi>g</mi><mi>e</mi><mi>d</mi><mi>a</mi><mi>t</mi><mi>a</mi></mrow><annotation encoding="application/x-tex">combinedFoldChange
data</annotation></semantics></math>gene &lt;- factor(data<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>g</mi><mi>e</mi><mi>n</mi><mi>e</mi><mo>,</mo><mi>l</mi><mi>e</mi><mi>v</mi><mi>e</mi><mi>l</mi><mi>s</mi><mo>=</mo><mi>u</mi><mi>n</mi><mi>i</mi><mi>q</mi><mi>u</mi><mi>e</mi><mo stretchy="false" form="prefix">(</mo><mi>d</mi><mi>a</mi><mi>t</mi><mi>a</mi></mrow><annotation encoding="application/x-tex">gene, levels = unique(data</annotation></semantics></math>gene))</p>
</div>
</div>
</div>
<div class="section level1">
<h1 id="keep-only-the-first-words-in-contrast-column-to-be-used-as-the-x-axis-labels">Keep only the first words in ‚Äòcontrast‚Äô column to be used as the x-axis labels.<a class="anchor" aria-label="anchor" href="#keep-only-the-first-words-in-contrast-column-to-be-used-as-the-x-axis-labels"></a>
</h1>
<p>data<span class="math inline">$contrast &lt;- sub(" .*", "", data$</span>contrast)</p>
</div>
<div class="section level1">
<h1 id="converting-the-contrast-column-as-factor-and-fix-the-current-level-order">Converting the ‚Äòcontrast‚Äô column as factor and fix the current level order<a class="anchor" aria-label="anchor" href="#converting-the-contrast-column-as-factor-and-fix-the-current-level-order"></a>
</h1>
<p>data<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>c</mi><mi>o</mi><mi>n</mi><mi>t</mi><mi>r</mi><mi>a</mi><mi>s</mi><mi>t</mi><mo>&lt;</mo><mi>‚àí</mi><mi>f</mi><mi>a</mi><mi>c</mi><mi>t</mi><mi>o</mi><mi>r</mi><mo stretchy="false" form="prefix">(</mo><mi>d</mi><mi>a</mi><mi>t</mi><mi>a</mi></mrow><annotation encoding="application/x-tex">contrast &lt;- factor(data</annotation></semantics></math>contrast, levels = unique(data$contrast))</p>
<p>p &lt;- plotFactor( data = data, x_col = ‚Äúcontrast‚Äù, y_col = ‚ÄúRE‚Äù, group_col = ‚Äúcontrast‚Äù, facet_col = ‚Äúgene‚Äù, Lower.se_col = ‚ÄúLower.se.RE‚Äù, Upper.se_col = ‚ÄúUpper.se.RE‚Äù, letters_col = ‚Äúsig‚Äù, letters_d = 0.2, alpha = 1, fill_colors = palette.colors(4, recycle = TRUE), color = ‚Äúblack‚Äù, col_width = 0.5, dodge_width = 0.5, base_size = 16, legend_position = ‚Äúnone‚Äù)</p>
<p>library(ggplot2) p + theme( panel.border = element_rect(color = ‚Äúblack‚Äù, linewidth = 0.5)) + theme(axis.text.x = element_text(size = 14, color = ‚Äúblack‚Äù, angle = 45, hjust = 1)) + xlab(NULL) + scale_y_continuous(expand = expansion(mult = c(0, 0.1))) + theme( strip.background = element_blank(), # removes the faceting gray background strip.text = element_text(face = ‚Äúbold‚Äù)) # optional: keeps the text visible</p>
<pre><code>&lt;p align="center"&gt;
&lt;img src="inst/Rplot03.png" width="80%"&gt;
&lt;/p&gt;

&lt;details&gt;
![Workflow of the analysis](figures/Rplot03.png)
&lt;/details&gt;


# Apply expression analysis per level of upper factor
When the dataset contains two factor columns, and the expression analysis is intended to be performed on the second factor within each level of the first factor, this can be achieved by looping over the levels of the first factor, as demonstrated in the following example.

&lt;details&gt;
&lt;summary&gt;&lt;strong&gt;See details&lt;/strong&gt;&lt;/summary&gt;

```{r eval= F}
library(rtpcr)
df &lt;- read.csv("E:/Dropbox/rtpcr manuscript/rtpcr/inst/extdata/farokh_et_al_2025.csv")
df


stage   stress  Rep TraesCS2D02G351900  CtTraesCS2D02G351900    TraesCS4A02G268500  CtTraesCS4A02G268500    TraesCS4B02G155800  CtTraesCS4B02G155800    TraesCS5B02G163500  CtTraesCS5B02G163500    TraesCS6A02G218900  CtTraesCS6A02G218900    TraesCS6B02G315400  CtTraesCS6B02G315400    TraesCS7A02G218600  CtTraesCS7A02G218600    TraesCS7B02G400600  CtTraesCS7B02G400600    EReference  CtRefrence
4LS Control 1   1.8 25.17   1.8 28.53   1.8 16.42   1.82    19.26   1.8 25.89   1.8 27.97   1.8 23.68   1.8 27.09   2   24.49
4LS Control 2   1.8 25.07   1.8 28.41   1.8 16.48   1.82    19.38   1.8 25.66   1.8 27.81   1.8 23.90   1.8 27.05   2   24.80
4LS Drought 1   1.8 26.72   1.8 28.08   1.8 15.60   1.82    20.07   1.8 23.6    1.8 26.77   1.8 26.21   1.8 27.58   2   25.55
4LS Drought 2   1.8 26.90   1.8 28.35   1.8 15.56   1.82    20.22   1.8 23.64   1.8 26.67   1.8 26.58   1.8 27.57   2   25.36
5LS Control 1   1.8 26.21   1.8 27.19   1.8 14.38   1.82    20.25   1.8 25.39   1.8 28.08   1.8 27.19   1.8 27.10   2   25.00
5LS Control 2   1.8 26.08   1.8 27.20   1.8 14.14   1.82    20.31   1.8 25.52   1.8 28.04   1.8 27.57   1.8 27.02   2   25.20
5LS Drought 1   1.8 26.78   1.8 25.75   1.8 13.23   1.82    20.07   1.8 24.84   1.8 26.38   1.8 27.07   1.8 27.69   2   26.05
5LS Drought 2   1.8 26.94   1.8 25.68   1.8 13.05   1.82    20.27   1.8 24.72   1.8 26.48   1.8 27.24   1.8 27.43   2   25.87
6LS Control 1   1.8 26.43   1.8 28.68   1.8 13.06   1.82    22.41   1.8 24.95   1.8 29.57   1.8 29.54   1.8 27.51   2   25.57
6LS Control 2   1.8 25.94   1.8 28.55   1.8 13.22   1.82    22.25   1.8 24.76   1.8 29.50   1.8 29.33   1.8 27.57   2   25.03
6LS Drought 1   1.8 25.78   1.8 25.56   1.8 11.87   1.82    20.93   1.8 24.67   1.8 27.88   1.8 28.43   1.8 28.29   2   26.44
6LS Drought 2   1.8 26.19   1.8 25.55   1.8 11.69   1.82    21.18   1.8 24.60   1.8 27.92   1.8 28.81   1.8 28.24   2   26.62
7LS Control 1   1.8 26.42   1.8 27.21   1.8 12.96   1.82    21.06   1.8 25.08   1.8 27.90   1.8 28.08   1.8 28.42   2   25.16
7LS Control 2   1.8 26.27   1.8 27.33   1.8 12.62   1.82    21.17   1.8 24.72   1.8 27.58   1.8 27.50   1.8 28.21   2   25.05
7LS Drought 1   1.8 25.75   1.8 26.00   1.8 11.35   1.82    18.78   1.8 23.92   1.8 27.00   1.8 28.21   1.8 28.27   2   25.57
7LS Drought 2   1.8 25.55   1.8 25.87   1.8 11.28   1.82    18.60   1.8 24.09   1.8 27.20   1.8 27.88   1.8 28.15   2   25.57



# Split by levels of the first column
df_split &lt;- split(df, df[[1]])

# Apply ANOVA_DDCt per level
res_list &lt;- lapply(names(df_split), function(lv) {

  df &lt;- df_split[[lv]]

  out &lt;- ANOVA_DDCt(
    df[, -1, drop = FALSE],
    numOfFactors = 1,
    numberOfrefGenes = 1,
    mainFactor.column = 1,
    analysisType = "anova",
    mainFactor.level.order = NULL,
    block = NULL
  )

  # Add level name as FIRST column
  out &lt;- cbind(
    Level = lv,
    out$combinedFoldChange
  )

  out
})

# Name list elements
names(res_list) &lt;- names(df_split)

# Combine all tables
final_table_1 &lt;- do.call(rbind, res_list)
rownames(final_table_1) &lt;- NULL</code></pre>

</div>
<div class="section level1">
<h1 id="post-hoc-analysis">Post-hoc analysis<a class="anchor" aria-label="anchor" href="#post-hoc-analysis"></a>
</h1>
<p>The <code>Means_DDCt</code> function performs post-hoc comparisons using a fitted model object produced by <code>ANOVA_DDCt</code> or <code>REPEATED_DDCt</code>. It applies pairwise statistical comparisons of relative expression (RE) values for user-specified effects via the <code>specs</code> argument. Supported effects include simple effects, interactions, and slicing, provided the underlying model is an ANOVA. For ANCOVA models returned by this package, the <code>Means_DDCt</code> output is limited to simple effects only.</p>
<p>```{r eval= F} res &lt;- ANOVA_DDCt( data_3factor, numOfFactors = 3, numberOfrefGenes = 1, mainFactor.column = 1, block = NULL)</p>
</div>
<div class="section level1">
<h1 id="relative-expression-values-for-concentration-main-effect">Relative expression values for Concentration main effect<a class="anchor" aria-label="anchor" href="#relative-expression-values-for-concentration-main-effect"></a>
</h1>
<p>Means_DDCt(res<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mi>e</mi><mi>r</mi><mi>G</mi><mi>e</mi><mi>n</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">perGene</annotation></semantics></math>E_PO$lm_ANOVA, specs = ‚ÄúConc‚Äù)</p>
<p>contrast RE SE df LCL UCL p.value sig L vs H 0.1703610 0.2208988 24 0.1242014 0.2336757 &lt;0.0001 <strong><em> M vs H 0.2227247 0.2208988 24 0.1623772 0.3055004 &lt;0.0001 </em></strong> M vs L 1.3073692 0.2208988 24 0.9531359 1.7932535 0.0928 .</p>
<p>Results are averaged over the levels of: Type, SA Confidence level used: 0.95</p>
</div>
<div class="section level1">
<h1 id="relative-expression-values-for-concentration-sliced-by-type">Relative expression values for Concentration sliced by Type<a class="anchor" aria-label="anchor" href="#relative-expression-values-for-concentration-sliced-by-type"></a>
</h1>
<p>Means_DDCt(res<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mi>e</mi><mi>r</mi><mi>G</mi><mi>e</mi><mi>n</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">perGene</annotation></semantics></math>E_PO$lm_ANOVA, specs = ‚ÄúConc | Type‚Äù)</p>
<p>Type = R: contrast RE SE df LCL UCL p.value sig L vs H 0.103187 0.3123981 24 0.0659984 0.161331 &lt;0.0001 <strong><em> M vs H 0.339151 0.3123981 24 0.2169210 0.530255 &lt;0.0001 </em></strong> M vs L 3.286761 0.3123981 24 2.1022126 5.138776 &lt;0.0001 ***</p>
<p>Type = S: contrast RE SE df LCL UCL p.value sig L vs H 0.281265 0.3123981 24 0.1798969 0.439751 &lt;0.0001 <strong><em> M vs H 0.146266 0.3123981 24 0.0935518 0.228684 &lt;0.0001 </em></strong> M vs L 0.520030 0.3123981 24 0.3326112 0.813055 0.0059 **</p>
<p>Results are averaged over the levels of: SA Confidence level used: 0.95</p>
</div>
<div class="section level1">
<h1 id="relative-expression-values-for-concentration-sliced-by-type-and-sa">Relative expression values for Concentration sliced by Type and SA<a class="anchor" aria-label="anchor" href="#relative-expression-values-for-concentration-sliced-by-type-and-sa"></a>
</h1>
<p>Means_DDCt(res<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mi>e</mi><mi>r</mi><mi>G</mi><mi>e</mi><mi>n</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">perGene</annotation></semantics></math>E_PO$lm_ANOVA, specs = ‚ÄúConc | Type * SA‚Äù)</p>
<pre><code>
# Checking normality of residuals

If the residuals from a `t.test` or an `lm` or and `lmer` object are not normally distributed, the significance results might be violated. In such cases, non-parametric tests can be used. For example, the Mann‚ÄìWhitney test (also known as the Wilcoxon rank-sum test), implemented via `wilcox.test()`, is an alternative to t.test, and `kruskal.test()` is an alternative to one-way analysis of variance. These tests assess differences between population medians using independent samples. However, the `t.test` function (along with the `TTEST_DDCt` function described above) includes the `var.equal` argument. When set to `FALSE`, perform `t.test` under the unequal variances hypothesis. Residuals for `lm` (from `ANOVA_DCt` and `ANOVA_DDCt` functions) and `lmer` (from `REPEATED_DDCt` function) objects can be extracted and plotted as follow:

```{r eval= F}
data &lt;- read.csv(system.file("extdata", "data_repeated_measure_1.csv", package = "rtpcr"))
res3 &lt;- REPEATED_DDCt(
  data,
  numOfFactors = 1,
  numberOfrefGenes = 1,
  repeatedFactor = "time",
  calibratorLevel = "1",
  block = NULL
)
residuals &lt;- resid(res3$perGene$Target$lm)
shapiro.test(residuals)
par(mfrow = c(1,2))
plot(residuals)
qqnorm(residuals)
qqline(residuals, col = "red")</code></pre>
</div>
<div class="section level1">
<h1 id="mean-of-technical-replicates">Mean of technical replicates<a class="anchor" aria-label="anchor" href="#mean-of-technical-replicates"></a>
</h1>
<p>The <code>meanTech</code> function calculates the mean of technical replicates by collapsing them into a single value, producing an output dataset that is ready for subsequent analyses. The input data must follow the column structure illustrated in the example below. Columns used to define technical replicate groupings should be specified using the <code>groups</code> argument of the <code>meanTech</code> function.</p>
<p>```{r eval= F} # See example input data frame: data &lt;- read.csv(system.file(‚Äúextdata‚Äù, ‚Äúdata_withTechRep.csv‚Äù, package = ‚Äúrtpcr‚Äù)) data</p>
</div>
<div class="section level1">
<h1 id="calculating-mean-of-technical-replicates">Calculating mean of technical replicates<a class="anchor" aria-label="anchor" href="#calculating-mean-of-technical-replicates"></a>
</h1>
<p>meanTech(data, groups = 1:4)</p>
<pre><code>





# Contact
Email: gh.mirzaghaderi at uok.ac.ir

# Citation
```md
citation("rtpcr")

To cite the package ‚Äòrtpcr‚Äô in publications, please use:

  Ghader Mirzaghaderi (2025). rtpcr: a package for statistical analysis and graphical
  presentation of qPCR data in R. PeerJ 13:e20185. https://doi.org/10.7717/peerj.20185

A BibTeX entry for LaTeX users is

  @Article{,
    title = {rtpcr: A package for statistical analysis and graphical presentation of qPCR data in R},
    author = {Ghader Mirzaghaderi},
    journal = {PeerJ},
    volume = {13},
    pages = {e20185},
    year = {2025},
    doi = {10.7717/peerj.20185},
  }</code></pre>
</div>
<div class="section level1">
<h1 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h1>
<p>Livak, Kenneth J, and Thomas D Schmittgen. 2001. Analysis of Relative Gene Expression Data Using Real-Time Quantitative PCR and the Double Delta CT Method. Methods 25 (4). <a href="https://doi.org/10.1006/meth.2001.1262" class="external-link">doi.org/10.1006/meth.2001.1262</a>.</p>
<p>Ganger, MT, Dietz GD, Ewing SJ. 2017. A common base method for analysis of qPCR data and the application of simple blocking in qPCR experiments. BMC bioinformatics 18, 1-11. <a href="https://doi.org/10.1186/s12859-017-1949-5" class="external-link">doi.org/10.1186/s12859-017-1949-5</a>.</p>
<p>Mirzaghaderi G. 2025. rtpcr: a package for statistical analysis and graphical presentation of qPCR data in R. PeerJ 13, e20185. <a href="https://doi.org/10.7717/peerj.20185" class="external-link">doi.org/10.7717/peerj.20185</a>.</p>
<p>Pfaffl MW, Horgan GW, Dempfle L. 2002. Relative expression software tool (REST¬©) for group-wise comparison and statistical analysis of relative expression results in real-time PCR. Nucleic acids research 30, e36-e36. <a href="https://doi.org/10.1093/nar/30.9.e36" class="external-link">doi.org/10.1093/nar/30.9.e36</a>.</p>
<p>Taylor SC, Nadeau K, Abbasi M, Lachance C, Nguyen M, Fenrich, J. 2019. The ultimate qPCR experiment: producing publication quality, reproducible data the first time. Trends in Biotechnology, 37(7), 761-774. <a href="https://doi.org/10.1016/j.tibtech.2018.12.002" class="external-link">doi.org/10.1016/j.tibtech.2018.12.002</a>.</p>
<p>Yuan, JS, Ann Reed, Feng Chen, and Neal Stewart. 2006. Statistical Analysis of Real-Time PCR Data. BMC Bioinformatics 7 (85). <a href="https://doi.org/10.1186/1471-2105-7-85" class="external-link">doi.org/10.1186/1471-2105-7-85</a>.</p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=rtpcr" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/mirzaghaderi/rtpcr/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing rtpcr</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Ghader Mirzaghaderi <br><small class="roles"> Author, maintainer, copyright holder </small>   </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://cran.rstudio.com/web/packages/rtpcr/index.html" class="external-link"><img src="https://cranlogs.r-pkg.org/badges/grand-total/rtpcr"></a></li>
<li><a href="https://github.com/mirzaghaderi/rtpcr" class="external-link"><img src="https://img.shields.io/github/r-package/v/mirzaghaderi/rtpcr" alt="R package version"></a></li>
<li><a href="https://CRAN.R-project.org/package=rtpcr" class="external-link"><img src="https://www.r-pkg.org/badges/version/rtpcr" alt="CRAN version"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Ghader Mirzaghaderi.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
