<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Calculation method • rtpcr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculation method">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rtpcr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/method.html">Calculation method</a></li>
    <li><a class="dropdown-item" href="../articles/vignette.html">rtpcr package help</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mirzaghaderi/rtpcr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Calculation method</h1>
                        <h4 data-toc-skip class="author">Ghader
Mirzaghaderi</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mirzaghaderi/rtpcr/blob/HEAD/vignettes/method.Rmd" class="external-link"><code>vignettes/method.Rmd</code></a></small>
      <div class="d-none name"><code>method.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="calculation-method">Calculation method<a class="anchor" aria-label="anchor" href="#calculation-method"></a>
</h2>
<p>Real-time polymerase chain reaction (real-time PCR) is widely used in
biological studies. Various analysis methods are employed on the
real-time PCR data to measure the mRNA levels under different
experimental conditions.</p>
<p>‘rtpcr’ package was developed for amplification efficiency
calculation, statistical analysis and bar plot representation of
real-time PCR data in R. By accounting for up to two reference genes and
amplification efficiency values, a general calculation methodology
described by <a href="https://doi.org/10.1186/s12859-017-1949-5" class="external-link">Ganger
et al. (2017)</a> and
<a href="https://doi.org/10.1016/j.tibtech.2018.12.002" class="external-link">Taylor et
al. (2019)</a>, matching both
<a href="https://doi.org/10.1006/meth.2001.1262" class="external-link">Livak and Schmittgen
(2001)</a> and <a href="https://doi.org/10.1093/nar/30.9.e36" class="external-link">Pfaffl et
al. (2002) </a> methods was used. Based on the experimental conditions,
the functions of the ‘rtpcr’ package use t-test (for experiments with a
two-level factor), analysis of variance, analysis of covariance (ANCOVA)
or analysis of repeated measure data to calculate the fold change (FC,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Δ</mi><mi>Δ</mi><msub><mi>C</mi><mi>t</mi></msub></mrow><annotation encoding="application/x-tex">{\Delta\Delta C_t}</annotation></semantics></math>
method) or relative expression (RE,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Δ</mi><msub><mi>C</mi><mi>t</mi></msub></mrow><annotation encoding="application/x-tex">{\Delta C_t}</annotation></semantics></math>
method). The functions further provide standard errors and confidence
interval for means, apply statistical mean comparisons and present
significance. To facilitate function application, different data sets
were used as examples and the outputs were explained. An outstanding
feature of ‘rtpcr’ package is providing publication-ready bar plots with
various controlling arguments for experiments with up to three different
factors which are further editable by ggplot2 functions.</p>
</div>
<div class="section level2">
<h2 id="calculations">Calculations<a class="anchor" aria-label="anchor" href="#calculations"></a>
</h2>
<p>The basic method for estimating gene expression between conditions
relies on calculating fold differences using the PCR amplification
efficiency
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>E</mi><annotation encoding="application/x-tex">E</annotation></semantics></math>)
and the threshold cycle (also called the crossing point,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>C</mi><mi>t</mi></msub><annotation encoding="application/x-tex">C_t</annotation></semantics></math>).
Among the various approaches for analyzing real-time PCR data, the Livak
method—also known as the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mn>2</mn><mrow><mi>−</mi><mi>Δ</mi><mi>Δ</mi><msub><mi>C</mi><mi>t</mi></msub></mrow></msup><annotation encoding="application/x-tex">2^{-\Delta\Delta C_t}</annotation></semantics></math>
method—is widely used for its simplicity and reliability. In this
method, the fold change (FC) expression,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mn>2</mn><mrow><mi>−</mi><mi>Δ</mi><mi>Δ</mi><msub><mi>C</mi><mi>t</mi></msub></mrow></msup><annotation encoding="application/x-tex">2^{-\Delta\Delta C_t}</annotation></semantics></math>,
in a Treatment (Tr) group relative to a Control (Co) group is calculated
as follows:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mtext mathvariant="normal">Fold change</mtext></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><msup><mn>2</mn><mrow><mi>−</mi><mi>Δ</mi><mi>Δ</mi><msub><mi>C</mi><mi>t</mi></msub></mrow></msup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mfrac><msup><mn>2</mn><mrow><mi>−</mi><msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>C</mi><mrow><mi>t</mi><mo>,</mo><mtext mathvariant="normal">target</mtext></mrow></msub><mo>−</mo><msub><mi>C</mi><mrow><mi>t</mi><mo>,</mo><mtext mathvariant="normal">ref</mtext></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mtext mathvariant="normal">Tr</mtext></msub></mrow></msup><msup><mn>2</mn><mrow><mi>−</mi><msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>C</mi><mrow><mi>t</mi><mo>,</mo><mtext mathvariant="normal">target</mtext></mrow></msub><mo>−</mo><msub><mi>C</mi><mrow><mi>t</mi><mo>,</mo><mtext mathvariant="normal">ref</mtext></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mtext mathvariant="normal">Co</mtext></msub></mrow></msup></mfrac></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><msup><mn>2</mn><mrow><mi>−</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>C</mi><mrow><mi>t</mi><mo>,</mo><mtext mathvariant="normal">target</mtext></mrow></msub><mo>−</mo><msub><mi>C</mi><mrow><mi>t</mi><mo>,</mo><mtext mathvariant="normal">ref</mtext></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mtext mathvariant="normal">Tr</mtext></msub><mo>−</mo><msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>C</mi><mrow><mi>t</mi><mo>,</mo><mtext mathvariant="normal">target</mtext></mrow></msub><mo>−</mo><msub><mi>C</mi><mrow><mi>t</mi><mo>,</mo><mtext mathvariant="normal">ref</mtext></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mtext mathvariant="normal">Co</mtext></msub><mo stretchy="true" form="postfix">]</mo></mrow></mrow></msup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><msup><mn>2</mn><mrow><mi>−</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>Δ</mi><msub><mi>C</mi><mi>t</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mtext mathvariant="normal">Tr</mtext></msub><mo>−</mo><msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>Δ</mi><msub><mi>C</mi><mi>t</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mtext mathvariant="normal">Co</mtext></msub><mo stretchy="true" form="postfix">]</mo></mrow></mrow></msup></mtd></mtr></mtable><annotation encoding="application/x-tex">
  \begin{align*}
\text{Fold change}
&amp;= 2^{-\Delta\Delta C_t} \\
&amp;= \frac{2^{-(C_{t,\text{target}} - C_{t,\text{ref}})_{\text{Tr}}}}
{2^{-(C_{t,\text{target}} - C_{t,\text{ref}})_{\text{Co}}}} \\
&amp;= 2^{-[(C_{t,\text{target}} - C_{t,\text{ref}})_{\text{Tr}}
        - (C_{t,\text{target}} - C_{t,\text{ref}})_{\text{Co}}]} \\
&amp;= 2^{-[(\Delta C_t)_{\text{Tr}} - (\Delta C_t)_{\text{Co}}]}
\end{align*}
</annotation></semantics></math></p>
<p>Here,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Δ</mi><msub><mi>C</mi><mi>t</mi></msub></mrow><annotation encoding="application/x-tex">\Delta C_t</annotation></semantics></math>
is the difference between target Ct and reference Ct values for a given
sample. Livak method assumes that both the target and reference genes
are amplified with efficiencies close to 100%, allowing for the relative
quantification of gene expression levels.</p>
<p>On the other hand, the Pfaffl method offers a more flexible approach
by accounting for differences in amplification efficiencies between the
target and reference genes. This method adjusts the calculated
expression ratio by incorporating the specific amplification
efficiencies, thus providing a more accurate representation of the
relative gene expression levels.</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mtext mathvariant="normal">Fold change</mtext><mo>=</mo><mfrac><msubsup><mi>E</mi><mtext mathvariant="normal">target</mtext><mrow><mi>−</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>C</mi><mrow><mi>t</mi><mo>,</mo><mtext mathvariant="normal">Tr</mtext></mrow></msub><mo>−</mo><msub><mi>C</mi><mrow><mi>t</mi><mo>,</mo><mtext mathvariant="normal">Co</mtext></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msubsup><msubsup><mi>E</mi><mtext mathvariant="normal">ref</mtext><mrow><mi>−</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>C</mi><mrow><mi>t</mi><mo>,</mo><mtext mathvariant="normal">Tr</mtext></mrow></msub><mo>−</mo><msub><mi>C</mi><mrow><mi>t</mi><mo>,</mo><mtext mathvariant="normal">Co</mtext></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msubsup></mfrac></mrow><annotation encoding="application/x-tex">
  \text{Fold change} =
  \frac{E_{\text{target}}^{-(C_{t,\text{Tr}} - C_{t,\text{Co}})}}
{E_{\text{ref}}^{-(C_{t,\text{Tr}} - C_{t,\text{Co}})}}
</annotation></semantics></math></p>
</div>
<div class="section level2">
<h2 id="a-generalized-calculation-method">A generalized calculation method<a class="anchor" aria-label="anchor" href="#a-generalized-calculation-method"></a>
</h2>
<p>The <code>rtpcr</code> package was developed for the R environment in
the major operating systems. The package functions are mainly based on
the calculation of efficiency-weighted
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Δ</mi><msub><mi>C</mi><mi>t</mi></msub></mrow><annotation encoding="application/x-tex">\Delta C_t</annotation></semantics></math><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">(</mo><mi>w</mi><mi>Δ</mi><msub><mi>C</mi><mi>t</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><annotation encoding="application/x-tex">(w\Delta C_t)</annotation></semantics></math>
values from target and reference gene Ct (equation 3).
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>w</mi><mi>Δ</mi><msub><mi>C</mi><mi>t</mi></msub></mrow><annotation encoding="application/x-tex">w\Delta C_t</annotation></semantics></math>
values are weighted for the amplification efficiencies as described by
Ganger et al. (2017) except that log2 is used instead of log10:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>w</mi><mi>Δ</mi><msub><mi>C</mi><mi>t</mi></msub><mo>=</mo><msub><mo>log</mo><mn>2</mn></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>E</mi><mtext mathvariant="normal">target</mtext></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>⋅</mo><msub><mi>C</mi><mrow><mi>t</mi><mo>,</mo><mtext mathvariant="normal">target</mtext></mrow></msub><mo>−</mo><msub><mo>log</mo><mn>2</mn></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>E</mi><mtext mathvariant="normal">ref</mtext></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>⋅</mo><msub><mi>C</mi><mrow><mi>t</mi><mo>,</mo><mtext mathvariant="normal">ref</mtext></mrow></msub></mrow><annotation encoding="application/x-tex">
  w\Delta C_t =
  \log_2(E_{\text{target}}) \cdot C_{t,\text{target}}
-
  \log_2(E_{\text{ref}}) \cdot C_{t,\text{ref}}
</annotation></semantics></math></p>
<p>The relative expression of the target gene normalized to that of
reference gene(s) within the same sample or condition is called relative
expression (RE). From the mean
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>w</mi><mi>Δ</mi><msub><mi>C</mi><mi>t</mi></msub></mrow><annotation encoding="application/x-tex">w\Delta C_t</annotation></semantics></math>
values over biological replicates, RE of a target gene can be calculated
for each condition according to the equation</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mtext mathvariant="normal">Relative Expression</mtext><mo>=</mo><msup><mn>2</mn><mrow><mi>−</mi><mover><mrow><mi>w</mi><mi>Δ</mi><msub><mi>C</mi><mi>t</mi></msub></mrow><mo accent="true">¯</mo></mover></mrow></msup></mrow><annotation encoding="application/x-tex">
  \text{Relative Expression} = 2^{-\overline{w\Delta C_t}}
</annotation></semantics></math></p>
<p>Relative expression is only calibrated for the reference gene(s) and
not for a control condition. However, often one condition is considered
as calibrator and the fold change (FC) expression in other conditions is
calculated relative to the calibrator. Examples are Treatment versus
Control where Control is served as the calibrator, or time 0 versus time
1 (e.g. after 1 hour) and time 2 (e.g. after 2 hours) where time 0 is
served as the reference or calibrator level. So, calibrator is the
reference level or sample that all others are compared to. The fold
change (FC) expression of a target gene for the reference or calibrator
level is 1 because it is not changed compared to itself. The fold change
expression of a target gene due to the treatment can be calculated as
follows:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mtext mathvariant="normal">Fold Change due to Treatment</mtext><mo>=</mo><msup><mn>2</mn><mrow><mi>−</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mover><mrow><mi>w</mi><mi>Δ</mi><msub><mi>C</mi><mi>t</mi></msub></mrow><mo accent="true">¯</mo></mover><mtext mathvariant="normal">Tr</mtext></msub><mo>−</mo><msub><mover><mrow><mi>w</mi><mi>Δ</mi><msub><mi>C</mi><mi>t</mi></msub></mrow><mo accent="true">¯</mo></mover><mtext mathvariant="normal">Co</mtext></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msup></mrow><annotation encoding="application/x-tex">
  \text{Fold Change due to Treatment} =
  2^{-(\overline{w\Delta C_t}_{\text{Tr}} - \overline{w\Delta C_t}_{\text{Co}})}
</annotation></semantics></math></p>
<p>Standard error of the FC and RE means is calculated according to
<a href="https://doi.org/10.1016/j.tibtech.2018.12.002" class="external-link">Taylor et
al. (2019)</a> in <code>rtpcr</code> package. Here, a brief methodology
is presented but details about the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>w</mi><mi>Δ</mi><msub><mi>C</mi><mi>t</mi></msub></mrow><annotation encoding="application/x-tex">w\Delta C_t</annotation></semantics></math>
calculations and statistical analysis are available in
<a href="https://doi.org/10.1186/s12859-017-1949-5" class="external-link">Ganger et
al. (2017)</a>. Importantly, because relative expression values follow a
lognormal distribution, a normal distribution is expected for the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>w</mi><mi>Δ</mi><msub><mi>C</mi><mi>t</mi></msub></mrow><annotation encoding="application/x-tex">w \Delta C_t</annotation></semantics></math>
values making it possible to apply t-test or analysis of variance.
Following analysis,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>w</mi><mi>Δ</mi><msub><mi>C</mi><mi>t</mi></msub></mrow><annotation encoding="application/x-tex">w\Delta C_t</annotation></semantics></math>
values are statistically compared and standard deviations and confidence
interval are calculated, but the transformation
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>y</mi><mo>=</mo><msup><mn>2</mn><mrow><mi>−</mi><mi>x</mi></mrow></msup></mrow><annotation encoding="application/x-tex">y = 2^{-x}</annotation></semantics></math>
is applied in the final step in order to report the results.</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Livak, Kenneth J, and Thomas D Schmittgen. 2001. Analysis of Relative
Gene Expression Data Using Real-Time Quantitative PCR and the Double
Delta CT Method. Methods 25 (4).
<a href="https://doi.org/10.1006/meth.2001.1262" class="external-link">doi.org/10.1006/meth.2001.1262</a>.</p>
<p>Ganger, MT, Dietz GD, Ewing SJ. 2017. A common base method for
analysis of qPCR data and the application of simple blocking in qPCR
experiments. BMC bioinformatics 18, 1-11.
<a href="https://doi.org/10.1186/s12859-017-1949-5" class="external-link">doi.org/10.1186/s12859-017-1949-5</a>.</p>
<p>Pfaffl MW, Horgan GW, Dempfle L. 2002. Relative expression software
tool (REST©) for group-wise comparison and statistical analysis of
relative expression results in real-time PCR. Nucleic acids research 30,
e36-e36.
<a href="https://doi.org/10.1093/nar/30.9.e36" class="external-link">doi.org/10.1093/nar/30.9.e36</a>.</p>
<p>Taylor SC, Nadeau K, Abbasi M, Lachance C, Nguyen M, Fenrich, J.
2019. The ultimate qPCR experiment: producing publication quality,
reproducible data the first time. Trends in Biotechnology, 37(7),
761-774.
<a href="https://doi.org/10.1016/j.tibtech.2018.12.002" class="external-link">doi.org/10.1016/j.tibtech.2018.12.002</a>.</p>
<p>Yuan, JS, Ann Reed, Feng Chen, and Neal Stewart. 2006. Statistical
Analysis of Real-Time PCR Data. BMC Bioinformatics 7 (85).
<a href="https://doi.org/10.1186/1471-2105-7-85" class="external-link">doi.org/10.1186/1471-2105-7-85</a>.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ghader Mirzaghaderi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
