<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Delta Delta Ct repeated measure analysis — REPEATED_DDCt • rtpcr</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Delta Delta Ct repeated measure analysis — REPEATED_DDCt"><meta name="description" content="REPEATED_DDCt function performs Delta Delta Ct (ddCt) method
analysis of observations repeatedly taken over different time courses.
Data may be obtained over time from a uni- or multi-factorial experiment."><meta property="og:description" content="REPEATED_DDCt function performs Delta Delta Ct (ddCt) method
analysis of observations repeatedly taken over different time courses.
Data may be obtained over time from a uni- or multi-factorial experiment."><meta property="og:image" content="https://mirzaghaderi.github.io/rtpcr/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rtpcr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Manual.html">Manual</a></li>
    <li><a class="dropdown-item" href="../articles/Method.html">Calculation method</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mirzaghaderi/rtpcr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Delta Delta Ct repeated measure analysis</h1>
      <small class="dont-index">Source: <a href="https://github.com/mirzaghaderi/rtpcr/blob/HEAD/R/REPEATED_DDCt.R" class="external-link"><code>R/REPEATED_DDCt.R</code></a></small>
      <div class="d-none name"><code>REPEATED_DDCt.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>REPEATED_DDCt</code> function performs Delta Delta Ct (ddCt) method
analysis of observations repeatedly taken over different time courses.
Data may be obtained over time from a uni- or multi-factorial experiment.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">REPEATED_DDCt</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">numOfFactors</span>,</span>
<span>  <span class="va">numberOfrefGenes</span>,</span>
<span>  <span class="va">mainFactor.column</span>,</span>
<span>  <span class="va">block</span>,</span>
<span>  mainFactor.level.order <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p.adj <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  analyseAllTarget <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>The input data frame containing experimental design columns, target gene
E/Ct column pairs, and reference gene E/Ct column pairs. Reference gene
columns must be located at the end of the data frame.</p></dd>


<dt id="arg-numoffactors">numOfFactors<a class="anchor" aria-label="anchor" href="#arg-numoffactors"></a></dt>
<dd><p>Integer. Number of experimental factor columns
(excluding <code>rep</code> and optional <code>block</code>).</p></dd>


<dt id="arg-numberofrefgenes">numberOfrefGenes<a class="anchor" aria-label="anchor" href="#arg-numberofrefgenes"></a></dt>
<dd><p>Integer. Number of reference genes.</p></dd>


<dt id="arg-mainfactor-column">mainFactor.column<a class="anchor" aria-label="anchor" href="#arg-mainfactor-column"></a></dt>
<dd><p>Integer. Column index of the factor (commonly <code>"time"</code>) for which relative expression is calculated.</p></dd>


<dt id="arg-block">block<a class="anchor" aria-label="anchor" href="#arg-block"></a></dt>
<dd><p>Character or <code>NULL</code>. Name of the blocking factor column.
When a qPCR experiment is done in multiple qPCR plates,
variation resulting from the plates may interfere with the actual amount of
gene expression. One solution is to conduct each plate as a randomized block
so that at least one replicate of each treatment and control is present
on a plate. Block effect is usually considered as random and its interaction
with any main effect is not considered.</p></dd>


<dt id="arg-mainfactor-level-order">mainFactor.level.order<a class="anchor" aria-label="anchor" href="#arg-mainfactor-level-order"></a></dt>
<dd><p>Optional character vector specifying the order of levels for the main factor.
If <code>NULL</code>, the first observed level is used as the calibrator.
If provided, the first element of the vector is used as the calibrator level.</p></dd>


<dt id="arg-p-adj">p.adj<a class="anchor" aria-label="anchor" href="#arg-p-adj"></a></dt>
<dd><p>Method for p-value adjustment. See <code><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></code>.</p></dd>


<dt id="arg-analysealltarget">analyseAllTarget<a class="anchor" aria-label="anchor" href="#arg-analysealltarget"></a></dt>
<dd><p>Logical or character.
If <code>TRUE</code> (default), all target genes are analysed.
Alternatively, a character vector specifying the names (names of their Efficiency columns) of target genes
to be analysed.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>An object containing expression table, lm model, residuals, raw data and ANOVA table for each gene:</p><dl><dt>ddCt expression table along with per-gene statistical comparison outputs</dt>
<dd><p><code>object$relativeExpression</code></p></dd>

<dt>ANOVA table</dt>
<dd><p><code>object$perGene$gene_name$ANOVA_table</code></p></dd>

<dt>lm ANOVA</dt>
<dd><p><code>object$perGene$gene_name$lm</code></p></dd>

<dt>lm_formula</dt>
<dd><p><code>object$perGene$gene_name$lm_formula</code></p></dd>

<dt>Residuals</dt>
<dd><p><code>resid(object$perGene$gene_name$lm)</code></p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>ddCt analysis of repeated measure data is performed for
the <code>mainFactor.column</code> based on a full model factorial experiment.</p>
<p>All the functions for relative expression analysis (including `TTEST_DDCt()`,
`WILCOX_DDCt()`, `ANOVA_DDCt()`, `ANCOVA_DDCt()`, `REPEATED_DDCt()`, and `ANOVA_DCt()`) return the
relative expression table which include fold change and corresponding
statistics. The output of `ANOVA_DDCt()`, `ANCOVA_DDCt()`, `ANCOVA_DDCt()`, `REPEATED_DDCt()`,
and `ANOVA_DCt()` also include lm models, residuals, raw data and ANOVA table
for each gene.</p>
<p>The expression table returned by `TTEST_DDCt()`,
`WILCOX_DDCt()`, `ANOVA_DDCt()`, `ANCOVA_DDCt()`, and `REPEATED_DDCt()` functions
include these columns: gene (name of target genes),
contrast (calibrator level and contrasts for which the relative expression is
computed), ddCt (mean of weighted delta delta Ct values), RE (relative expression or
fold change = 2^-ddCt),  log2FC (log(2) of relative expression or fold change),
pvalue, sig (per-gene significance), LCL (95% lower confidence level), UCL (95% upper confidence level),
se (standard error of mean calculated from the weighted delta Ct values of each of the main factor levels),
Lower.se.RE (The lower limit error bar for RE which is 2^(log2(RE) - se)),
Upper.se.RE (The upper limit error bar for RE which is 2^(log2(RE) + se)),
Lower.se.log2FC (The lower limit error bar for log2 RE), and
Upper.se.log2FC (The upper limit error bar for log2 RE)</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>LivakKJ, Schmittgen TD (2001).
Analysis of Relative Gene Expression Data Using Real-Time Quantitative PCR
and the Double Delta CT Method.
<em>Methods</em>, 25(4), 402–408.
doi:10.1006/meth.2001.1262</p>
<p>Ganger MT, Dietz GD, and Ewing SJ (2017).
A common base method for analysis of qPCR data and the application of
simple blocking in qPCR experiments.
<em>BMC Bioinformatics</em>, 18, 1–11.</p>
<p>Taylor SC, Nadeau K, Abbasi M, Lachance C, Nguyen M, Fenrich, J. (2019).
The ultimate qPCR experiment: producing publication quality, reproducible
data the first time. <em>Trends in Biotechnology</em>, 37, 761-774.</p>
<p>Yuan JS, Reed A, Chen F, Stewart N (2006).
Statistical Analysis of Real-Time PCR Data.
<em>BMC Bioinformatics</em>, 7, 85.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">data1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"data_repeated_measure_1.csv"</span>, package <span class="op">=</span> <span class="st">"rtpcr"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">REPEATED_DDCt</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">data1</span>,</span></span>
<span class="r-in"><span>  numOfFactors <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>  numberOfrefGenes <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>  mainFactor.column <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>  block <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Relative Expression</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     gene       contrast     ddCt      RE   log2FC  pvalue sig    LCL      UCL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 Target              1  0.00000 1.00000  0.00000 1.00000     0.0000  0.00000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Target time2 vs time1  0.22333 0.85658 -0.22333 0.81660     0.0923  7.94918</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Target time3 vs time1 -2.23333 4.70219  2.23333 0.06847   . 0.5067 43.63677</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        se Lower.se.RE Upper.se.RE Lower.se.log2FC Upper.se.log2FC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 1.40507     0.37760     2.64830         0.00000         0.00000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 0.97527     0.43570     1.68405        -0.43907        -0.11360</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 0.55411     3.20256     6.90403         1.52108         3.27911</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The 1 level was used as calibrator.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">data2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"data_repeated_measure_2.csv"</span>, package <span class="op">=</span> <span class="st">"rtpcr"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">REPEATED_DDCt</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">data2</span>,</span></span>
<span class="r-in"><span>  numOfFactors <span class="op">=</span> <span class="fl">2</span>,</span></span>
<span class="r-in"><span>  numberOfrefGenes <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>  mainFactor.column <span class="op">=</span> <span class="fl">2</span>, </span></span>
<span class="r-in"><span>  block <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  p.adj <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Relative Expression</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     gene       contrast     ddCt      RE  log2FC  pvalue sig     LCL     UCL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 Target              1  0.00000 1.00000 0.00000 1.00000     0.00000 0.00000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Target time2 vs time1 -0.32833 1.25556 0.32833 0.55402     0.43805 3.59873</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Target time3 vs time1 -1.34667 2.54324 1.34667 0.03509   * 0.88731 7.28951</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        se Lower.se.RE Upper.se.RE Lower.se.log2FC Upper.se.log2FC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 0.70359     0.61404     1.62855         0.00000         0.00000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 0.53234     0.86814     1.81588         0.22702         0.47486</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 0.70742     1.55752     4.15279         0.82472         2.19894</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The 1 level was used as calibrator.</span>
<span class="r-in"><span>          </span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ghader Mirzaghaderi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

