<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>\(\Delta\Delta C_T\) method t-test analysis — TTEST_DDCt • rtpcr</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="\(\Delta\Delta C_T\) method t-test analysis — TTEST_DDCt"><meta property="og:description" content="The TTEST_DDCt function performs fold change expression analysis based on
the \(\Delta\Delta C_T\) method using Student's t-test. It supports analysis
of one or more target genes evaluated under two experimental conditions
(e.g. control vs treatment)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rtpcr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mirzaghaderi/rtpcr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>\(\Delta\Delta C_T\) method t-test analysis</h1>
    <small class="dont-index">Source: <a href="https://github.com/mirzaghaderi/rtpcr/blob/HEAD/R/TTEST_DDCt.r" class="external-link"><code>R/TTEST_DDCt.r</code></a></small>
    <div class="hidden name"><code>TTEST_DDCt.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The <code>TTEST_DDCt</code> function performs fold change expression analysis based on
the \(\Delta\Delta C_T\) method using Student's t-test. It supports analysis
of one or more target genes evaluated under two experimental conditions
(e.g. control vs treatment).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">TTEST_DDCt</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">numberOfrefGenes</span>,</span>
<span>  Factor.level.order <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  paired <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  var.equal <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  p.adj <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  order <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  plotType <span class="op">=</span> <span class="st">"RE"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A data frame containing experimental conditions, biological replicates, and
amplification efficiency and Ct values for target and reference genes.
The number of biological replicates must be equal across genes. If this
is not true, or there are <code>NA</code> values use <code>ANODA_DDCt</code> function
for independent samples or <code>REPEATED_DDCt</code> for paired samples.
See the package vignette for details on the required data structure.</p></dd>


<dt id="arg-numberofrefgenes">numberOfrefGenes<a class="anchor" aria-label="anchor" href="#arg-numberofrefgenes"></a></dt>
<dd><p>Integer specifying the number of reference genes used for normalization
(must be \(\ge 1\)).</p></dd>


<dt id="arg-factor-level-order">Factor.level.order<a class="anchor" aria-label="anchor" href="#arg-factor-level-order"></a></dt>
<dd><p>Optional character vector specifying the order of factor levels.
If <code>NULL</code>, the first level of the factor column is used as the calibrator.</p></dd>


<dt id="arg-paired">paired<a class="anchor" aria-label="anchor" href="#arg-paired"></a></dt>
<dd><p>Logical; if <code>TRUE</code>, a paired t-test is performed.</p></dd>


<dt id="arg-var-equal">var.equal<a class="anchor" aria-label="anchor" href="#arg-var-equal"></a></dt>
<dd><p>Logical; if <code>TRUE</code>, equal variances are assumed and a pooled variance
estimate is used. Otherwise, Welch's t-test is applied.</p></dd>


<dt id="arg-p-adj">p.adj<a class="anchor" aria-label="anchor" href="#arg-p-adj"></a></dt>
<dd><p>Method for p-value adjustment. One of
<code>"holm"</code>, <code>"hochberg"</code>, <code>"hommel"</code>, <code>"bonferroni"</code>,
<code>"BH"</code>, <code>"BY"</code>, <code>"fdr"</code>, or <code>"none"</code>.</p></dd>


<dt id="arg-order">order<a class="anchor" aria-label="anchor" href="#arg-order"></a></dt>
<dd><p>Optional character vector specifying the order of genes in the output plot.</p></dd>


<dt id="arg-plottype">plotType<a class="anchor" aria-label="anchor" href="#arg-plottype"></a></dt>
<dd><p>Plot scale to use: <code>"RE"</code> for relative expression or
<code>"log2FC"</code> for log2 fold change.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A list with the following components:</p><dl><dt>Result</dt>
<dd><p>Table containing RE values, log2FC, p-values, significance codes,
  confidence intervals, standard errors, and lower/upper SE limits.</p></dd>

  <dt>RE_Plot</dt>
<dd><p>Bar plot of relative expression values.</p></dd>

  <dt>log2FC_Plot</dt>
<dd><p>Bar plot of log2 fold change values.</p></dd>


</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>Relative expression values are computed using one or more reference genes for normalization.
Both paired and unpaired experimental designs are supported.</p>
<p>Paired samples in quantitative PCR refer to measurements collected from the same
individuals under two different conditions (e.g. before vs after treatment),
whereas unpaired samples originate from different individuals in each condition.
Paired designs allow within-individual comparisons and typically reduce
inter-individual variability.</p>
<p>The function returns numerical summaries as well as bar plots based on either
relative expression (RE) or log2 fold change (log2FC).</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Livak, K. J. and Schmittgen, T. D. (2001).
Analysis of Relative Gene Expression Data Using Real-Time Quantitative PCR
and the Double Delta CT Method.
<em>Methods</em>, 25(4), 402–408.
doi:10.1006/meth.2001.1262</p>
<p>Ganger, M. T., Dietz, G. D., and Ewing, S. J. (2017).
A common base method for analysis of qPCR data and the application of
simple blocking in qPCR experiments.
<em>BMC Bioinformatics</em>, 18, 1–11.</p>
<p>Yuan, J. S., Reed, A., Chen, F., and Stewart, N. (2006).
Statistical Analysis of Real-Time PCR Data.
<em>BMC Bioinformatics</em>, 7, 85.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Ghader Mirzaghaderi</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Example data structure</span></span></span>
<span class="r-in"><span><span class="va">data1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"data_1factor_one_ref.csv"</span>, package <span class="op">=</span> <span class="st">"rtpcr"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Unpaired t-test</span></span></span>
<span class="r-in"><span><span class="fu">TTEST_DDCt</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">data1</span>,</span></span>
<span class="r-in"><span>  paired <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>  var.equal <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  numberOfrefGenes <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> *** 3 target(s) using 1 reference gene(s) was analysed!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> *** The control level was used as calibrator.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Result</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      Gene     RE    LCL    UCL pvalue     se  log2FC Lower.se.RE Upper.se.RE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 C2H2_26 0.4373 0.1926 0.9927 0.0488 0.4218 -1.1933      0.3264      0.5858</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 C2H2_01 4.0185 2.4598 6.5649 0.0014 0.2193  2.0067      3.4518      4.6783</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 C2H2_12 1.6472 0.9595 2.8279 0.0624 0.2113  0.7200      1.4228      1.9069</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   sig Lower.se.log2FC Upper.se.log2FC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   *         -1.5985         -0.8908</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  **          1.7237          2.3361</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3              0.6219          0.8335</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $plot</span>
<span class="r-plt img"><img src="TTEST_DDCt-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># With amplification efficiencies</span></span></span>
<span class="r-in"><span><span class="va">data2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"data_1factor_one_ref_Eff.csv"</span>, package <span class="op">=</span> <span class="st">"rtpcr"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">TTEST_DDCt</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">data2</span>,</span></span>
<span class="r-in"><span>  paired <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>  var.equal <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  numberOfrefGenes <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> *** 1 target(s) using 1 reference gene(s) was analysed!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> *** The control level was used as calibrator.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Result</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Gene    RE    LCL    UCL pvalue     se  log2FC Lower.se.RE Upper.se.RE sig</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 target 0.409 0.2388 0.7006 0.0099 0.1985 -1.2897      0.3565      0.4694  **</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Lower.se.log2FC Upper.se.log2FC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1           -1.48          -1.124</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $plot</span>
<span class="r-plt img"><img src="TTEST_DDCt-2.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Two reference genes</span></span></span>
<span class="r-in"><span><span class="va">data3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"data_1factor_Two_ref.csv"</span>, package <span class="op">=</span> <span class="st">"rtpcr"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">TTEST_DDCt</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">data3</span>,</span></span>
<span class="r-in"><span>  numberOfrefGenes <span class="op">=</span> <span class="fl">2</span>,</span></span>
<span class="r-in"><span>  var.equal <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  p.adj <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> *** 1 target(s) using 2 reference gene(s) was analysed!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> *** The control level was used as calibrator.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Result</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Gene     RE    LCL    UCL pvalue     se  log2FC Lower.se.RE Upper.se.RE sig</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 DER5 0.4343 0.1921 0.9816 0.0469 0.4191 -1.2033      0.3248      0.5807   *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Lower.se.log2FC Upper.se.log2FC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1          -1.609            -0.9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $plot</span>
<span class="r-plt img"><img src="TTEST_DDCt-3.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ghader Mirzaghaderi.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer></div>






  </body></html>

